

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="English" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="English" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Advanced User Guide - SangerContig (AB1) &mdash; sangeranalyseR  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Advanced User Guide - SangerAlignment (AB1)" href="advance_user_SangerAlignment_ab1.html" />
    <link rel="prev" title="Advanced User Guide - SangerRead (AB1)" href="advance_user_SangerRead_ab1.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> sangeranalyseR
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Beginners Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerRead_ab1.html">Advanced User Guide - <em>SangerRead</em> (<strong>AB1</strong>)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Advanced User Guide - <em>SangerContig</em> (<strong>AB1</strong>)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#preparing-sangercontig-ab1-inputs">Preparing <em>SangerContig</em> <strong>AB1</strong> inputs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#regular-expression-matching-sangercontig-inputs-ab1">(1) &quot;regular expression matching&quot; <em>SangerContig</em> inputs (<strong>AB1</strong>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#csv-file-matching-sangercontig-inputs-ab1">(2) &quot;CSV file matching&quot; <em>SangerContig</em> inputs (<strong>AB1</strong>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#creating-sangercontig-instance-from-ab1">Creating <em>SangerContig</em> instance from <strong>AB1</strong></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#regular-expression-matching-sangercontig-creation-ab1">(1) &quot;regular expression matching&quot; <em>SangerContig</em> creation (<strong>AB1</strong>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#csv-file-matching-sangercontig-creation-ab1">(2) &quot;CSV file matching&quot; <em>SangerContig</em> creation (<strong>AB1</strong>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#updating-sangercontig-quality-trimming-parameters">Updating <em>SangerContig</em> quality trimming parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#launching-sangercontig-shiny-app">Launching <em>SangerContig</em> Shiny app</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sangercontig-page-sc-app"><em>SangerContig</em> page (SC app)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sangerread-page-sc-app"><em>SangerRead</em> page (SC app)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#writing-sangercontig-fasta-files-ab1">Writing <em>SangerContig</em> FASTA files <sub>(AB1)</sub></a></li>
<li class="toctree-l2"><a class="reference internal" href="#generating-sangercontig-report-ab1">Generating <em>SangerContig</em> report <sub>(AB1)</sub></a></li>
<li class="toctree-l2"><a class="reference internal" href="#code-summary-sangercontig-ab1">Code summary (<em>SangerContig</em>, <strong>AB1</strong>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id3">(1) Preparing <em>SangerContig</em> <strong>AB1</strong> inputs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">(2) Creating <em>SangerContig</em> instance from <strong>AB1</strong></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#regular-expression-method-sangercontig-creation-ab1">(2.1) &quot;Regular Expression Method&quot; <em>SangerContig</em> creation (<strong>AB1</strong>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id5">(2.2) &quot;CSV file matching&quot; <em>SangerContig</em> creation (<strong>AB1</strong>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id6">(3) Updating <em>SangerContig</em> quality trimming parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id7">(4) Launching <em>SangerContig</em> Shiny app</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id8">(5) Writing <em>SangerContig</em> FASTA files <sub>(AB1)</sub></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id9">(6) Generating <em>SangerContig</em> report <sub>(AB1)</sub></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerAlignment_ab1.html">Advanced User Guide - <em>SangerAlignment</em> (<strong>AB1</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerRead_fasta.html">Advanced User Guide - <em>SangerRead</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerContig_fasta.html">Advanced User Guide - <em>SangerContig</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerAlignment_fasta.html">Advanced User Guide - <em>SangerAlignment</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_page.html">Q &amp; A ...</a></li>
<li class="toctree-l1"><a class="reference internal" href="function_manual.html">User Manual (functions)</a></li>
<li class="toctree-l1"><a class="reference internal" href="frequent_questions.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="conclusion.html">Conclusion</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="contact.html">Contact</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Help</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">sangeranalyseR</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Advanced User Guide - <em>SangerContig</em> (<strong>AB1</strong>)</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/advance_user_SangerContig_ab1.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <section id="advanced-user-guide-sangercontig-ab1">
<h1>Advanced User Guide - <em>SangerContig</em> (<strong>AB1</strong>)<a class="headerlink" href="#advanced-user-guide-sangercontig-ab1" title="Permalink to this headline">¶</a></h1>
<p><em>SangerContig</em> is in the intermediate level of sangeranalyseR (<a class="reference internal" href="#sangercontig-hierarchy"><span class="std std-ref">Figure_1</span></a>), and each <em>SangerContig</em> instance corresponds to a contig in a Sanger sequencing experiment. Among its slots, there are two lists, forward and reverse read list, storing <em>SangerRead</em> in the corresponding direction.</p>
<p>In this section, we are going to go through details about a reproducible <em>SangerContig</em> analysis example with the <strong>AB1</strong> file input in sangeranalyseR. By running the following example codes, you will get an end-to-end <em>SangerContig</em> analysis result.</p>
<figure class="align-center" id="id10">
<span id="sangercontig-hierarchy"></span><a class="reference internal image-reference" href="../_images/SangerContig_hierarchy.png"><img alt="../_images/SangerContig_hierarchy.png" src="../_images/SangerContig_hierarchy.png" style="width: 586.6px; height: 217.8px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 1. Hierarchy of classes in sangeranalyseR, <em>SangerContig</em> level.</span><a class="headerlink" href="#id10" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<div class="line-block">
<div class="line"><br /></div>
</div>
<section id="preparing-sangercontig-ab1-inputs">
<h2>Preparing <em>SangerContig</em> <strong>AB1</strong> inputs<a class="headerlink" href="#preparing-sangercontig-ab1-inputs" title="Permalink to this headline">¶</a></h2>
<p>The main input file format to create <em>SangerContig</em> instance is <strong>AB1</strong>. Before starting the analysis, users need to prepare one directory containing all <strong>AB1</strong> files, and all of them must be in the first layer of that directory. In other words, there should be no subdirectories. In this example, the data are in the sangeranalyseR package; thus, you can simply get its path by running the following codes:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">rawDataDir</span> <span class="o">&lt;-</span> <span class="nf">system.file</span><span class="p">(</span><span class="s">&quot;extdata&quot;</span><span class="p">,</span> <span class="n">package</span> <span class="o">=</span> <span class="s">&quot;sangeranalyseR&quot;</span><span class="p">)</span>
<span class="n">parentDir</span> <span class="o">&lt;-</span> <span class="nf">file.path</span><span class="p">(</span><span class="n">rawDataDir</span><span class="p">,</span> <span class="s">&quot;Allolobophora_chlorotica&quot;</span><span class="p">,</span> <span class="s">&quot;RBNII&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The value of <code class="code docutils literal notranslate"><span class="pre">parentDir</span></code> is where all <strong>AB1</strong> files are placed. If your operating system is macOS, then its value should look like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">Library</span><span class="o">/</span><span class="n">Frameworks</span><span class="o">/</span><span class="n">R</span><span class="o">.</span><span class="n">framework</span><span class="o">/</span><span class="n">Versions</span><span class="o">/</span><span class="mf">4.0</span><span class="o">/</span><span class="n">Resources</span><span class="o">/</span><span class="n">library</span><span class="o">/</span><span class="n">sangeranalyseR</span><span class="o">/</span><span class="n">extdata</span><span class="o">/</span><span class="n">Allolobophora_chlorotica</span><span class="o">/</span><span class="n">RBNII</span>
</pre></div>
</div>
<p>And we showed the files under <code class="code docutils literal notranslate"><span class="pre">parentDir</span></code> in <a class="reference internal" href="#sangercontig-file-structure-ab1"><span class="std std-ref">Figure_2</span></a>:</p>
<figure class="align-center" id="id11">
<span id="sangercontig-file-structure-ab1"></span><a class="reference internal image-reference" href="../_images/SangerContig_file_structure.png"><img alt="../_images/SangerContig_file_structure.png" src="../_images/SangerContig_file_structure.png" style="width: 276.0px; height: 390.0px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 2. <em>SangerContig</em> filename regulation.</span><a class="headerlink" href="#id11" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p><a class="reference internal" href="#sangercontig-file-structure-ab1"><span class="std std-ref">Figure_2</span></a> shows the file-naming regulation and hierarchy. In this example, <code class="code docutils literal notranslate"><span class="pre">RBNII</span></code> is the parent directory, and all <strong>AB1</strong> files must be under its first layer. There are two ways for users to group their <strong>AB1</strong> files which are <strong>&quot;regular expression matching&quot;</strong> and <strong>&quot;CSV file matching&quot;</strong>, and following are instructions of how to prepare and name your <strong>AB1</strong> input files.</p>
<section id="regular-expression-matching-sangercontig-inputs-ab1">
<h3>(1) &quot;regular expression matching&quot; <em>SangerContig</em> inputs (<strong>AB1</strong>)<a class="headerlink" href="#regular-expression-matching-sangercontig-inputs-ab1" title="Permalink to this headline">¶</a></h3>
<p>For regular expression matching method, sangeranalyseR will group <strong>AB1</strong> files based on their contig names and read directions in their filenames automatically; therefore, users have to follow the file-naming regulations below:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>All input files must have <strong>.ab1</strong> as its file extension</p></li>
<li><p>All input files must have the same contig name in their filenames.</p></li>
<li><p>Forward or reverse direction has to be specified in the filename.</p></li>
</ul>
</div>
<p>There are four parameters, <code class="code docutils literal notranslate"><span class="pre">ABIF_Directory</span></code>, <code class="code docutils literal notranslate"><span class="pre">contigName</span></code>, <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixForward</span></code>, and <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixReverse</span></code>, that define the grouping rule to let sangeranalyseR automatically match correct <strong>AB1</strong> files and divide them into forward and reverse directions.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">ABIF_Directory</span></code>: this is the directory that contains all <strong>AB1</strong> files, and it can be either an absolute or relative path. We suggest users to put only target <strong>AB1</strong> files inside this directory and do not include any other unrelated files.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">contigName</span></code>: this is a regular expression that matches filenames that are going to be included in the <em>SangerContig</em> analysis. <code class="code docutils literal notranslate"><span class="pre">grepl</span></code> function in R is used.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixForward</span></code>: this is a regular expression that matches all filenames in forward direction. <code class="code docutils literal notranslate"><span class="pre">grepl</span></code> function in R is used.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixReverse</span></code>: this is a regular expression that matches all filenames in reverse direction. <code class="code docutils literal notranslate"><span class="pre">grepl</span></code> function in R is used.</p></li>
</ul>
</div>
<p>If you don't know what regular expression is, don't panic - it's just a way of recognising text. Please refer to <a class="reference internal" href="how_to_page.html#what-is-a-regular-expression"><span class="std std-ref">What is a regular expression?</span></a> for more details. Here is an example of how it works in sangeranalseR:</p>
<p>So how sangeranalyseR works is that it first matches the <code class="code docutils literal notranslate"><span class="pre">contigName</span></code> to exclude unrelated files and then separate the forward and reverse reads by matching <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixForward</span></code> and <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixReverse</span></code>. Therefore, it is important to make sure that all target <strong>AB1</strong> files share the same <code class="code docutils literal notranslate"><span class="pre">contigName</span></code> and carefully select your <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixForward</span></code> and <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixReverse</span></code>. The bad file-naming and wrong regex matching might accidentally include reverse reads into the forward read list or vice versa, which will make the program generate wrong results. Therefore, it is important to have a consistent naming strategy. So, how should we systematically name <strong>AB1</strong> files? We suggest users to follow the file-naming regulation in <a class="reference internal" href="#sangeranalyser-filename-convention-sangercontig"><span class="std std-ref">Figure_3</span></a>.</p>
<figure class="align-center" id="id12">
<span id="sangeranalyser-filename-convention-sangercontig"></span><a class="reference internal image-reference" href="../_images/sangeranalyseR_filename_convention.png"><img alt="../_images/sangeranalyseR_filename_convention.png" src="../_images/sangeranalyseR_filename_convention.png" style="width: 621.0px; height: 28.0px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 3. Suggested <strong>AB1</strong> file-naming regulation - <em>SangerContig</em>.</span><a class="headerlink" href="#id12" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>As you can see, the first part of the regulation is a consensus read name (or contig name), which narrows down the scope of <strong>AB1</strong> files to those we are going to examine. The second part of the regulation is an index. Since there might be more than one read that is in the forward or reverse direction, we recommend you to number your reads in the same contig group. The third part is a direction which is either 'F' (forward) or 'R' (reverse). Last but not least, files have to end with <strong>.ab1</strong> file extension.</p>
<p>To make it more specific, let's go back to the true example. In <a class="reference internal" href="#sangercontig-file-structure-ab1"><span class="std std-ref">Figure_2</span></a>, there are a lot of <strong>AB1</strong> files from different contigs in <code class="code docutils literal notranslate"><span class="pre">RBNII</span></code> (<code class="code docutils literal notranslate"><span class="pre">ABIF_Directory</span></code>).
First, we set <code class="code docutils literal notranslate"><span class="pre">contigName</span></code> to <code class="code docutils literal notranslate"><span class="pre">&quot;Achl_RBNII384-13&quot;</span></code> to reduce candidates from eight to two <strong>AB1</strong> files, <code class="code docutils literal notranslate"><span class="pre">Achl_RBNII384-13_1_F.ab1</span></code> and <code class="code docutils literal notranslate"><span class="pre">Achl_RBNII384-13_2_R.ab1</span></code>. Then, we set <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixForward</span></code> to <code class="code docutils literal notranslate"><span class="pre">&quot;_[0-9]*_F.ab1$&quot;</span></code> and <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixReverse</span></code> to <code class="code docutils literal notranslate"><span class="pre">&quot;_[0-9]*_R.ab1$&quot;</span></code> to let sangeranalyseR match and group forward and reverse reads automatically. By the regular expression rule, <code class="code docutils literal notranslate"><span class="pre">Achl_RBNII384-13_1_F.ab1</span></code> and <code class="code docutils literal notranslate"><span class="pre">Achl_RBNII384-13_2_R.ab1</span></code> will be categorized into &quot;forward read list&quot; and &quot;reverse read list&quot; respectively. The reason why we strongly recommend you to follow this file-naming regulation is that by doing so, you can directly adopt the example regular expression matching values, <code class="code docutils literal notranslate"><span class="pre">&quot;_[0-9]*_F.ab1$&quot;</span></code> and <code class="code docutils literal notranslate"><span class="pre">&quot;_[0-9]*_R.ab1$&quot;</span></code>, to group reads and reduce chances of error.</p>
<p>After understanding how parameters work, please refer to <a class="reference internal" href="#creating-sangercontig-instance-from-ab1"><span class="std std-ref">Creating SangerContig instance from AB1</span></a> below to see how sangeranalseR creates 'Achl_RBNII384-13' <em>SangerContig</em> instance.</p>
</section>
<section id="csv-file-matching-sangercontig-inputs-ab1">
<h3>(2) &quot;CSV file matching&quot; <em>SangerContig</em> inputs (<strong>AB1</strong>)<a class="headerlink" href="#csv-file-matching-sangercontig-inputs-ab1" title="Permalink to this headline">¶</a></h3>
<p>For those who are not familiar with regular expression, we provide a second grouping approach, CSV file matching method. sangeranalyseR will group <strong>AB1</strong> files based on the information in a CSV file automatically; therefore, users have to follow the regulations below:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Here is an <a class="reference download internal" download="" href="../_downloads/5650fc7db022343113057d999c117b2d/names_conversion_2.csv"><code class="xref download docutils literal notranslate"><span class="pre">example</span> <span class="pre">CSV</span> <span class="pre">file</span></code></a> (<a class="reference internal" href="#sangeranalyser-csv-file-sangercontig-ab1"><span class="std std-ref">Figure_4</span></a>)</p>
<blockquote>
<div><figure class="align-center" id="id13">
<span id="sangeranalyser-csv-file-sangercontig-ab1"></span><a class="reference internal image-reference" href="../_images/sangeranalyseR_csv_file_sangercontig_ab1.png"><img alt="../_images/sangeranalyseR_csv_file_sangercontig_ab1.png" src="../_images/sangeranalyseR_csv_file_sangercontig_ab1.png" style="width: 598.5px; height: 72.1px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 4. Example CSV file for <em>SangerContig</em> instance creation.</span><a class="headerlink" href="#id13" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</div></blockquote>
<ul class="simple">
<li><p>There must be three columns, &quot;<strong>reads</strong>&quot;, &quot;<strong>direction</strong>&quot;, and &quot;<strong>contig</strong>&quot;, in the CSV file.</p></li>
<li><p>The &quot;<strong>reads</strong>&quot; column stores the filename of <strong>AB1</strong> files that are going to be included in the analysis.</p></li>
<li><p>The &quot;<strong>direction</strong>&quot; column stores the direction of the reads. It must be &quot;F&quot; (forward) or &quot;R&quot; (reverse).</p></li>
<li><p>The &quot;<strong>contig</strong>&quot; column stores the contig name that each read blongs. Reads in the same contig have to have the same contig name, and they will be grouped into the same <em>SangerContig</em> instance.</p></li>
</ul>
</div>
<p>There are three parameters, <code class="code docutils literal notranslate"><span class="pre">ABIF_Directory</span></code>, <code class="code docutils literal notranslate"><span class="pre">contigName</span></code>, and <code class="code docutils literal notranslate"><span class="pre">CSV_NamesConversion</span></code>,that define the grouping rule to help sangeranalseR to automatically match correct <strong>AB1</strong> files and divide them into forward and reverse directions.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">ABIF_Directory</span></code>: this is the directory that contains all <strong>AB1</strong> files, and it can be either an absolute or relative path. We suggest users to put only target AB1 files inside this directory and do not include any other unrelated files.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">contigName</span></code>: this is a regular expression that matches filenames that are going to be included in the <em>SangerContig</em> analysis. <code class="code docutils literal notranslate"><span class="pre">grepl</span></code> function in R is used.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">CSV_NamesConversion</span></code>: this is the path to the CSV file. It can be either an absolute or relative path.</p></li>
</ul>
</div>
<p>The main difference between &quot;CSV file matching&quot; and &quot;regular expression matching&quot; is where the grouping rule is written. For &quot;regular expression matching&quot;, rules are writtein in filenames, and thus more naming requirements are required. In contrast, rules of &quot;CSV file matching&quot; are written in an additional CSV file so it is more flexible on <strong>AB1</strong> file-naming.</p>
<p>So how sangeranalyseR works is that it first reads in the CSV file (with <em>&quot;reads&quot;</em>, <em>&quot;direction&quot;</em>, and <em>&quot;contig&quot;</em> columns), filter out rows whose <em>&quot;contig&quot;</em> is not the value of <code class="code docutils literal notranslate"><span class="pre">contigName</span></code> parameter, find the names of <strong>AB1</strong> files listed in <em>&quot;reads&quot;</em>, and assign directions to them based on <em>&quot;direction&quot;</em>.</p>
<p>To make it more specific, let's go back to the true example. First, we prepare a <a class="reference download internal" download="" href="../_downloads/5650fc7db022343113057d999c117b2d/names_conversion_2.csv"><code class="xref download docutils literal notranslate"><span class="pre">CSV</span> <span class="pre">file</span></code></a> (<code class="code docutils literal notranslate"><span class="pre">CSV_NamesConversion</span></code>) and a file directory like <a class="reference internal" href="#sangercontig-file-structure-ab1"><span class="std std-ref">Figure_2</span></a> (<code class="code docutils literal notranslate"><span class="pre">ABIF_Directory</span></code>) with some <strong>AB1</strong> files from different contigs. In the CSV file, both rows have the contig name <code class="code docutils literal notranslate"><span class="pre">&quot;Achl_RBNII384-13&quot;</span></code>, which is what we need to assign to the <code class="code docutils literal notranslate"><span class="pre">contigName</span></code> parameter. sangeranalyseR then checks and matches <em>&quot;reads&quot;</em> of these two rows, <code class="code docutils literal notranslate"><span class="pre">&quot;Achl_RBNII384-13_1_F.ab1&quot;</span></code> and <code class="code docutils literal notranslate"><span class="pre">&quot;Achl_RBNII384-13_2_R.ab1&quot;</span></code>, in <code class="code docutils literal notranslate"><span class="pre">RBNII</span></code> directory and reduce candidates from eight to two <strong>AB1</strong> files. Last, these two reads are assigned into &quot;forward read list&quot; and &quot;reverse read list&quot; respectively by the <em>&quot;direction&quot;</em> column.</p>
<p>After understanding how parameters work, please refer to <a class="reference internal" href="#creating-sangercontig-instance-from-ab1"><span class="std std-ref">Creating SangerContig instance from AB1</span></a> below to see how sangeranalseR creates 'Achl_RBNII384-13' <em>SangerContig</em> instance.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="creating-sangercontig-instance-from-ab1">
<h2>Creating <em>SangerContig</em> instance from <strong>AB1</strong><a class="headerlink" href="#creating-sangercontig-instance-from-ab1" title="Permalink to this headline">¶</a></h2>
<p>After preparing the input directory, we can create a <em>SangerContig</em> instance by running <code class="code docutils literal notranslate"><span class="pre">SangerContig</span></code> constructor function or <code class="code docutils literal notranslate"><span class="pre">new</span></code> method. The constructor function is a wrapper for <code class="code docutils literal notranslate"><span class="pre">new</span></code> method and it makes instance creation more intuitive. Their input parameters are same, and all of them have their default values. For more details about <em>SangerContig</em> inputs and slots definition, please refer to <a class="reference external" href="https://bioconductor.org/packages/release/bioc/manuals/sangeranalyseR/man/sangeranalyseR.pdf">sangeranalyseR reference manual</a>. We will explain two <em>SangerContig</em> instance creation methods, &quot;regular expression matching&quot; and &quot;CSV file matching&quot;.</p>
<section id="regular-expression-matching-sangercontig-creation-ab1">
<h3>(1) &quot;regular expression matching&quot; <em>SangerContig</em> creation (<strong>AB1</strong>)<a class="headerlink" href="#regular-expression-matching-sangercontig-creation-ab1" title="Permalink to this headline">¶</a></h3>
<p>The consturctor function and <code class="code docutils literal notranslate"><span class="pre">new</span></code> method below contain four parameters, <code class="code docutils literal notranslate"><span class="pre">ABIF_Directory</span></code>, <code class="code docutils literal notranslate"><span class="pre">contigName</span></code>, <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixForward</span></code>, and <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixReverse</span></code>, that we mentioned in the previous section. It also includes important parameters like quality trimming, chromatogram visualization, consensus alignment, and so on. Run the following code and create <code class="code docutils literal notranslate"><span class="pre">my_sangerContig</span></code> instance.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># using `constructor` function to create SangerContig instance</span>
<span class="n">my_sangerContig</span> <span class="o">&lt;-</span> <span class="nf">SangerContig</span><span class="p">(</span><span class="n">inputSource</span>           <span class="o">=</span> <span class="s">&quot;ABIF&quot;</span><span class="p">,</span>
                                <span class="n">processMethod</span>         <span class="o">=</span> <span class="s">&quot;REGEX&quot;</span><span class="p">,</span>
                                <span class="n">ABIF_Directory</span>        <span class="o">=</span> <span class="n">parentDir</span><span class="p">,</span>
                                <span class="n">contigName</span>            <span class="o">=</span> <span class="s">&quot;Achl_RBNII384-13&quot;</span><span class="p">,</span>
                                <span class="n">REGEX_SuffixForward</span>   <span class="o">=</span> <span class="s">&quot;_[0-9]*_F.ab1$&quot;</span><span class="p">,</span>
                                <span class="n">REGEX_SuffixReverse</span>   <span class="o">=</span> <span class="s">&quot;_[0-9]*_R.ab1$&quot;</span><span class="p">,</span>
                                <span class="n">TrimmingMethod</span>        <span class="o">=</span> <span class="s">&quot;M1&quot;</span><span class="p">,</span>
                                <span class="n">M1TrimmingCutoff</span>      <span class="o">=</span> <span class="m">0.0001</span><span class="p">,</span>
                                <span class="n">M2CutoffQualityScore</span>  <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                                <span class="n">M2SlidingWindowSize</span>   <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                                <span class="n">baseNumPerRow</span>         <span class="o">=</span> <span class="m">100</span><span class="p">,</span>
                                <span class="n">heightPerRow</span>          <span class="o">=</span> <span class="m">200</span><span class="p">,</span>
                                <span class="n">signalRatioCutoff</span>     <span class="o">=</span> <span class="m">0.33</span><span class="p">,</span>
                                <span class="n">showTrimmed</span>           <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                                <span class="n">refAminoAcidSeq</span>       <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">,</span>
                                <span class="n">minReadsNum</span>           <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
                                <span class="n">minReadLength</span>         <span class="o">=</span> <span class="m">20</span><span class="p">,</span>
                                <span class="n">minFractionCall</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                                <span class="n">maxFractionLost</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                                <span class="n">geneticCode</span>           <span class="o">=</span> <span class="n">GENETIC_CODE</span><span class="p">,</span>
                                <span class="n">acceptStopCodons</span>      <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                                <span class="n">readingFrame</span>          <span class="o">=</span> <span class="m">1</span><span class="p">,</span>
                                <span class="n">processorsNum</span>         <span class="o">=</span> <span class="m">1</span><span class="p">)</span>

<span class="c1"># using `new` method to create SangerContig instance</span>
<span class="n">my_sangerContig</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;SangerContig&quot;</span><span class="p">,</span>
                       <span class="n">inputSource</span>           <span class="o">=</span> <span class="s">&quot;ABIF&quot;</span><span class="p">,</span>
                       <span class="n">processMethod</span>         <span class="o">=</span> <span class="s">&quot;REGEX&quot;</span><span class="p">,</span>
                       <span class="n">ABIF_Directory</span>        <span class="o">=</span> <span class="n">parentDir</span><span class="p">,</span>
                       <span class="n">contigName</span>            <span class="o">=</span> <span class="s">&quot;Achl_RBNII384-13&quot;</span><span class="p">,</span>
                       <span class="n">REGEX_SuffixForward</span>   <span class="o">=</span> <span class="s">&quot;_[0-9]*_F.ab1$&quot;</span><span class="p">,</span>
                       <span class="n">REGEX_SuffixReverse</span>   <span class="o">=</span> <span class="s">&quot;_[0-9]*_R.ab1$&quot;</span><span class="p">,</span>
                       <span class="n">TrimmingMethod</span>        <span class="o">=</span> <span class="s">&quot;M1&quot;</span><span class="p">,</span>
                       <span class="n">M1TrimmingCutoff</span>      <span class="o">=</span> <span class="m">0.0001</span><span class="p">,</span>
                       <span class="n">M2CutoffQualityScore</span>  <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                       <span class="n">M2SlidingWindowSize</span>   <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                       <span class="n">baseNumPerRow</span>         <span class="o">=</span> <span class="m">100</span><span class="p">,</span>
                       <span class="n">heightPerRow</span>          <span class="o">=</span> <span class="m">200</span><span class="p">,</span>
                       <span class="n">signalRatioCutoff</span>     <span class="o">=</span> <span class="m">0.33</span><span class="p">,</span>
                       <span class="n">showTrimmed</span>           <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                       <span class="n">refAminoAcidSeq</span>       <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">,</span>
                       <span class="n">minReadsNum</span>           <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
                       <span class="n">minReadLength</span>         <span class="o">=</span> <span class="m">20</span><span class="p">,</span>
                       <span class="n">minFractionCall</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                       <span class="n">maxFractionLost</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                       <span class="n">geneticCode</span>           <span class="o">=</span> <span class="n">GENETIC_CODE</span><span class="p">,</span>
                       <span class="n">acceptStopCodons</span>      <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                       <span class="n">readingFrame</span>          <span class="o">=</span> <span class="m">1</span><span class="p">,</span>
                       <span class="n">processorsNum</span>         <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
<p>In this example, <code class="code docutils literal notranslate"><span class="pre">contigName</span></code> is set to <code class="code docutils literal notranslate"><span class="pre">Achl_RBNII384-13</span></code>, so only <code class="code docutils literal notranslate"><span class="pre">Achl_RBNII384-13_1_F.ab1</span></code> and <code class="code docutils literal notranslate"><span class="pre">Achl_RBNII384-13_2_R.ab1</span></code> are selected. Moreover, by regular expression pattern matching, <code class="code docutils literal notranslate"><span class="pre">Achl_RBNII384-13_1_F.ab1</span></code> is categorized into the forward list, and <code class="code docutils literal notranslate"><span class="pre">Achl_RBNII384-13_2_R.ab1</span></code> is categorized into the reverse read. Both reads are aligned into a contig, <code class="code docutils literal notranslate"><span class="pre">my_sangerContig</span></code>, and it will be used as the input for the following functions.</p>
<p>Inside the R shell, you can run <code class="code docutils literal notranslate"><span class="pre">my_sangerContig</span></code> to get basic information of the instance or run <code class="code docutils literal notranslate"><span class="pre">my_sangerContig&#64;objectResults&#64;readResultTable</span></code> to check the creation result of every Sanger read after <code class="code docutils literal notranslate"><span class="pre">my_sangerContig</span></code> is successfully created.</p>
<p>Here is the output of <code class="code docutils literal notranslate"><span class="pre">my_sangerContig</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>SangerContig S4 instance
         Input Source :  ABIF
         Process Method :  REGEX
         ABIF Directory :  /Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata/Allolobophora_chlorotica/RBNII
   REGEX Suffix Forward :  _[0-9]*_F.ab1$
   REGEX Suffix Reverse :  _[0-9]*_R.ab1$
            Contig Name :  Achl_RBNII384-13
         &#39;minReadsNum&#39; :  2
      &#39;minReadLength&#39; :  20
      &#39;minFractionCall&#39; :  0.5
      &#39;maxFractionLost&#39; :  0.5
   &#39;acceptStopCodons&#39; :  TRUE
         &#39;readingFrame&#39; :  1
      Contig Sequence :  AGCAGGATAGTAGGGGCTGGTATAAGACTCCTAATTCGAATTGAGCTAAGACAGCCGGGAGCATTTCTAGGAAGGGATCAACTCTATAACACTATTGTAACTGCTCACGCATTTGTAATAATTTTCTTTCTAGTAATACCTGTATTTATTGGGGGGTTCGGTAATTGACTTCTACCTTTAATACTTGGAGCCCCTGACATGGCATTCCCACGTCTTAACAACATAAGATTTTGACTCCTTCCCCCATCACTAATCCTTCTAGTATCCTCTGCTGCAGTAGAAAAGGGTGCGGGAACTGGATGAACTGTTTATCCACCCCTAGCAAGAAACATTGCTCATGCCGGCCCATCTGTAGACTTAGCTATTTTTTCTCTTCATTTAGCAGGTGCTTCATCAATCTTGGGTGCCATTAATTTTATTACTACTGTTATTAACATACGATGAAGAGGCTTACGACTTGAACGAATCCCATTATTCGTTTGAGCCGTACTAATTACAGTGGTCCTTCTACTCTTATCTTTACCAGTATTAGCCGGTGCAATTACTATACTACTTACCGATCGAAATCTAAATACCTCCTTCTTTGACCCTGCTGGAGGCGGAGAT
Forward reads in the contig &gt;&gt;  1
Reverse reads in the contig &gt;&gt;  1
SUCCESS [2021-12-07 17:01:18] &#39;Achl_RBNII384-13&#39; is successfully created!
</pre></div>
</div>
<p>Here is the output of <code class="code docutils literal notranslate"><span class="pre">my_sangerContig&#64;objectResults&#64;readResultTable</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>                  <span class="n">readName</span> <span class="n">creationResult</span> <span class="n">errorType</span> <span class="n">errorMessage</span> <span class="n">inputSource</span>    <span class="n">direction</span>
<span class="mi">1</span> <span class="n">Achl_RBNII384</span><span class="o">-</span><span class="mi">13_1</span><span class="n">_F</span><span class="o">.</span><span class="n">ab1</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>        <span class="n">ABIF</span> <span class="n">Forward</span> <span class="n">Read</span>
<span class="mi">2</span> <span class="n">Achl_RBNII384</span><span class="o">-</span><span class="mi">13_2</span><span class="n">_R</span><span class="o">.</span><span class="n">ab1</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>        <span class="n">ABIF</span> <span class="n">Reverse</span> <span class="n">Read</span>
</pre></div>
</div>
</section>
<section id="csv-file-matching-sangercontig-creation-ab1">
<h3>(2) &quot;CSV file matching&quot; <em>SangerContig</em> creation (<strong>AB1</strong>)<a class="headerlink" href="#csv-file-matching-sangercontig-creation-ab1" title="Permalink to this headline">¶</a></h3>
<p>The consturctor function and <code class="code docutils literal notranslate"><span class="pre">new</span></code> method below contain three parameters, <code class="code docutils literal notranslate"><span class="pre">ABIF_Directory</span></code>, <code class="code docutils literal notranslate"><span class="pre">contigName</span></code>, and <code class="code docutils literal notranslate"><span class="pre">CSV_NamesConversion</span></code>, that we mentioned in the previous section. It also includes important parameters like quality trimming, chromatogram visualization, consensus alignment, and so on. Run the following code and create <code class="code docutils literal notranslate"><span class="pre">my_sangerContig</span></code> instance.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">csv_namesConversion</span> <span class="o">&lt;-</span> <span class="nf">file.path</span><span class="p">(</span><span class="n">rawDataDir</span><span class="p">,</span> <span class="s">&quot;ab1&quot;</span><span class="p">,</span> <span class="s">&quot;SangerContig&quot;</span><span class="p">,</span> <span class="s">&quot;names_conversion_2.csv&quot;</span><span class="p">)</span>

<span class="c1"># using `constructor` function to create SangerContig instance</span>
<span class="n">my_sangerContig</span> <span class="o">&lt;-</span> <span class="nf">SangerContig</span><span class="p">(</span><span class="n">inputSource</span>            <span class="o">=</span> <span class="s">&quot;ABIF&quot;</span><span class="p">,</span>
                                <span class="n">processMethod</span>          <span class="o">=</span> <span class="s">&quot;CSV&quot;</span><span class="p">,</span>
                                <span class="n">ABIF_Directory</span>         <span class="o">=</span> <span class="n">parentDir</span><span class="p">,</span>
                                <span class="n">contigName</span>             <span class="o">=</span> <span class="s">&quot;Achl_RBNII384-13&quot;</span><span class="p">,</span>
                                <span class="n">CSV_NamesConversion</span>    <span class="o">=</span> <span class="n">csv_namesConversion</span><span class="p">,</span>
                                <span class="n">TrimmingMethod</span>         <span class="o">=</span> <span class="s">&quot;M1&quot;</span><span class="p">,</span>
                                <span class="n">M1TrimmingCutoff</span>       <span class="o">=</span> <span class="m">0.0001</span><span class="p">,</span>
                                <span class="n">M2CutoffQualityScore</span>   <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                                <span class="n">M2SlidingWindowSize</span>    <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                                <span class="n">baseNumPerRow</span>          <span class="o">=</span> <span class="m">100</span><span class="p">,</span>
                                <span class="n">heightPerRow</span>           <span class="o">=</span> <span class="m">200</span><span class="p">,</span>
                                <span class="n">signalRatioCutoff</span>      <span class="o">=</span> <span class="m">0.33</span><span class="p">,</span>
                                <span class="n">showTrimmed</span>            <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                                <span class="n">refAminoAcidSeq</span>        <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">,</span>
                                <span class="n">minReadsNum</span>            <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
                                <span class="n">minReadLength</span>          <span class="o">=</span> <span class="m">20</span><span class="p">,</span>
                                <span class="n">minFractionCall</span>        <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                                <span class="n">maxFractionLost</span>        <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                                <span class="n">geneticCode</span>            <span class="o">=</span> <span class="n">GENETIC_CODE</span><span class="p">,</span>
                                <span class="n">acceptStopCodons</span>       <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                                <span class="n">readingFrame</span>           <span class="o">=</span> <span class="m">1</span><span class="p">,</span>
                                <span class="n">processorsNum</span>          <span class="o">=</span> <span class="m">1</span><span class="p">)</span>


<span class="c1"># using `new` method to create SangerContig instance</span>
<span class="n">my_sangerContig</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;SangerContig&quot;</span><span class="p">,</span>
                       <span class="n">inputSource</span>           <span class="o">=</span> <span class="s">&quot;ABIF&quot;</span><span class="p">,</span>
                       <span class="n">processMethod</span>         <span class="o">=</span> <span class="s">&quot;CSV&quot;</span><span class="p">,</span>
                       <span class="n">ABIF_Directory</span>        <span class="o">=</span> <span class="n">parentDir</span><span class="p">,</span>
                       <span class="n">contigName</span>            <span class="o">=</span> <span class="s">&quot;Achl_RBNII384-13&quot;</span><span class="p">,</span>
                       <span class="n">CSV_NamesConversion</span>   <span class="o">=</span> <span class="n">csv_namesConversion</span><span class="p">,</span>
                       <span class="n">TrimmingMethod</span>         <span class="o">=</span> <span class="s">&quot;M1&quot;</span><span class="p">,</span>
                       <span class="n">M1TrimmingCutoff</span>       <span class="o">=</span> <span class="m">0.0001</span><span class="p">,</span>
                       <span class="n">M2CutoffQualityScore</span>   <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                       <span class="n">M2SlidingWindowSize</span>    <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                       <span class="n">baseNumPerRow</span>          <span class="o">=</span> <span class="m">100</span><span class="p">,</span>
                       <span class="n">heightPerRow</span>           <span class="o">=</span> <span class="m">200</span><span class="p">,</span>
                       <span class="n">signalRatioCutoff</span>      <span class="o">=</span> <span class="m">0.33</span><span class="p">,</span>
                       <span class="n">showTrimmed</span>            <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                       <span class="n">refAminoAcidSeq</span> <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">,</span>
                       <span class="n">minReadsNum</span>            <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
                       <span class="n">minReadLength</span>          <span class="o">=</span> <span class="m">20</span><span class="p">,</span>
                       <span class="n">minFractionCall</span>        <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                       <span class="n">maxFractionLost</span>        <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                       <span class="n">geneticCode</span>            <span class="o">=</span> <span class="n">GENETIC_CODE</span><span class="p">,</span>
                       <span class="n">acceptStopCodons</span>       <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                       <span class="n">readingFrame</span>           <span class="o">=</span> <span class="m">1</span><span class="p">,</span>
                       <span class="n">processorsNum</span>          <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
<p>First, you need to load the CSV file into the R environment. If you are still don't know how to prepare it, please check <a class="reference internal" href="#csv-file-matching-sangercontig-inputs-ab1"><span class="std std-ref">(2) &quot;CSV file matching&quot; SangerContig inputs (AB1)</span></a>. Then, it will follow rules in the CSV file and create <code class="code docutils literal notranslate"><span class="pre">my_sangerContig</span></code>. After it's created, inside the R shell, you can run <code class="code docutils literal notranslate"><span class="pre">my_sangerContig</span></code> to get basic information of the instance or run <code class="code docutils literal notranslate"><span class="pre">my_sangerContig&#64;objectResults&#64;readResultTable</span></code> to check the creation result of every Sanger read after <code class="code docutils literal notranslate"><span class="pre">my_sangerContig</span></code> is successfully created.</p>
<p>Here is the output of <code class="code docutils literal notranslate"><span class="pre">my_sangerContig</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>SangerContig S4 instance
         Input Source :  ABIF
         Process Method :  CSV
         ABIF Directory :  /Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata/Allolobophora_chlorotica/RBNII
   CSV Names Conversion :  /Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata/ab1/SangerContig/names_conversion_2.csv
            Contig Name :  Achl_RBNII384-13
         &#39;minReadsNum&#39; :  2
      &#39;minReadLength&#39; :  20
      &#39;minFractionCall&#39; :  0.5
      &#39;maxFractionLost&#39; :  0.5
   &#39;acceptStopCodons&#39; :  TRUE
         &#39;readingFrame&#39; :  1
      Contig Sequence :  AGCAGGATAGTAGGGGCTGGTATAAGACTCCTAATTCGAATTGAGCTAAGACAGCCGGGAGCATTTCTAGGAAGGGATCAACTCTATAACACTATTGTAACTGCTCACGCATTTGTAATAATTTTCTTTCTAGTAATACCTGTATTTATTGGGGGGTTCGGTAATTGACTTCTACCTTTAATACTTGGAGCCCCTGACATGGCATTCCCACGTCTTAACAACATAAGATTTTGACTCCTTCCCCCATCACTAATCCTTCTAGTATCCTCTGCTGCAGTAGAAAAGGGTGCGGGAACTGGATGAACTGTTTATCCACCCCTAGCAAGAAACATTGCTCATGCCGGCCCATCTGTAGACTTAGCTATTTTTTCTCTTCATTTAGCAGGTGCTTCATCAATCTTGGGTGCCATTAATTTTATTACTACTGTTATTAACATACGATGAAGAGGCTTACGACTTGAACGAATCCCATTATTCGTTTGAGCCGTACTAATTACAGTGGTCCTTCTACTCTTATCTTTACCAGTATTAGCCGGTGCAATTACTATACTACTTACCGATCGAAATCTAAATACCTCCTTCTTTGACCCTGCTGGAGGCGGAGAT
Forward reads in the contig &gt;&gt;  1
Reverse reads in the contig &gt;&gt;  1
SUCCESS [2021-12-07 17:11:48] &#39;Achl_RBNII384-13&#39; is successfully created!
</pre></div>
</div>
<p>Here is the output of <code class="code docutils literal notranslate"><span class="pre">my_sangerContig&#64;objectResults&#64;readResultTable</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>                  <span class="n">readName</span> <span class="n">creationResult</span> <span class="n">errorType</span> <span class="n">errorMessage</span> <span class="n">inputSource</span>    <span class="n">direction</span>
<span class="mi">1</span> <span class="n">Achl_RBNII384</span><span class="o">-</span><span class="mi">13_1</span><span class="n">_F</span><span class="o">.</span><span class="n">ab1</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>        <span class="n">ABIF</span> <span class="n">Forward</span> <span class="n">Read</span>
<span class="mi">2</span> <span class="n">Achl_RBNII384</span><span class="o">-</span><span class="mi">13_2</span><span class="n">_R</span><span class="o">.</span><span class="n">ab1</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>        <span class="n">ABIF</span> <span class="n">Reverse</span> <span class="n">Read</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="updating-sangercontig-quality-trimming-parameters">
<h2>Updating <em>SangerContig</em> quality trimming parameters<a class="headerlink" href="#updating-sangercontig-quality-trimming-parameters" title="Permalink to this headline">¶</a></h2>
<p>In the previous <a class="reference internal" href="#creating-sangercontig-instance-from-ab1"><span class="std std-ref">Creating SangerContig instance from AB1</span></a> part, the constructor function will apply the quality trimming parameters to all reads. After creating a <em>SangerContig</em> instance, users can change the trimming parameters by running <code class="code docutils literal notranslate"><span class="pre">updateQualityParam</span></code> function which will update all reads with the new trimming parameters and redo reads alignment. If users want to do quality trimming read by read instead of all at once, please move on to the next section, <a class="reference internal" href="#launching-sangercontig-shiny-app"><span class="std std-ref">Launching SangerContig Shiny app</span></a> page.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">newSangerContig</span> <span class="o">&lt;-</span> <span class="nf">updateQualityParam</span><span class="p">(</span><span class="n">my_sangerContig</span><span class="p">,</span>
                                      <span class="n">TrimmingMethod</span>       <span class="o">=</span> <span class="s">&quot;M2&quot;</span><span class="p">,</span>
                                      <span class="n">M1TrimmingCutoff</span>     <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                                      <span class="n">M2CutoffQualityScore</span> <span class="o">=</span> <span class="m">20</span><span class="p">,</span>
                                      <span class="n">M2SlidingWindowSize</span>  <span class="o">=</span> <span class="m">15</span><span class="p">)</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="launching-sangercontig-shiny-app">
<h2>Launching <em>SangerContig</em> Shiny app<a class="headerlink" href="#launching-sangercontig-shiny-app" title="Permalink to this headline">¶</a></h2>
<p>We create an interactive local Shiny app for users to go into each <em>SangerRead</em> in <em>SangerContig</em> instance. Users only need to run one function, <code class="code docutils literal notranslate"><span class="pre">launchApp</span></code>, with previously created instance as input and the <em>SangerContig</em> Shiny app will pop up. Here, we will go through <em>SangerRead</em> and <em>SangerContig</em> pages.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">launchApp</span><span class="p">(</span><span class="n">my_sangerContig</span><span class="p">)</span>
</pre></div>
</div>
<section id="sangercontig-page-sc-app">
<h3><em>SangerContig</em> page (SC app)<a class="headerlink" href="#sangercontig-page-sc-app" title="Permalink to this headline">¶</a></h3>
<p><em>SangerContig</em> page is the initial page of <em>SangerContig</em> Shiny app. <a class="reference internal" href="#sangercontig-shiny-sangercontig-page"><span class="std std-ref">Figure 5</span></a> shows the overview page of the contig. Notice that there is a red &quot;Re-calculate Contig&quot; button. Users need to click the button after changing the quality trimming parameters in order to get the updated information. In SangerContig page, there are two expendable tabs, “Forward Reads” and “Reverse Reads” storing the corresponding reads on the left-hand side navigation panel in <a class="reference internal" href="#sangercontig-shiny-sangercontig-page"><span class="std std-ref">Figure 5</span></a>. See <a class="reference internal" href="#sangerread-page-sc-app"><span class="std std-ref">SangerRead page (SC app)</span></a> for more details of the subpage.</p>
<figure class="align-center" id="id14">
<span id="sangercontig-shiny-sangercontig-page"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_SangerContig_page.png"><img alt="../_images/SangerContig_shiny_SangerContig_page.png" src="../_images/SangerContig_shiny_SangerContig_page.png" style="width: 576.0px; height: 318.0px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 5. <em>SangerContig</em> Shiny app initial page - <em>SangerContig</em> page.</span><a class="headerlink" href="#id14" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>The information provided in this page are input parameters and contig results including “genetic code table”, “reference amino acid sequence”, “reads alignment”, “difference data frame”, “dendrogram”, “sample distance heatmap”, “indels data frame”, and “stop codons data frame”.</p>
<p><a class="reference internal" href="#sangercontig-shiny-alignment-differencedf"><span class="std std-ref">Figure 6</span></a> shows reads alignment result and difference data frame. The alignment is generated by <code class="code docutils literal notranslate"><span class="pre">AlignSeqs</span></code> or <code class="code docutils literal notranslate"><span class="pre">AlignTranslation</span></code> function in <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/DECIPHER.html">DECIPHER</a> package.</p>
<figure class="align-center" id="id15">
<span id="sangercontig-shiny-alignment-differencedf"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_alignment_differenceDF.png"><img alt="../_images/SangerContig_shiny_alignment_differenceDF.png" src="../_images/SangerContig_shiny_alignment_differenceDF.png" style="width: 660.0px; height: 361.2px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 6. <em>SangerContig</em> page - reads alignment and difference data frame.</span><a class="headerlink" href="#id15" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p><a class="reference internal" href="#sangercontig-shiny-dendrogram"><span class="std std-ref">Figure 7</span></a> shows dendrogram result in both plot and in data frame. The results are generated by <code class="code docutils literal notranslate"><span class="pre">TreeLine</span></code> function in <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/DECIPHER.html">DECIPHER</a> package.</p>
<figure class="align-center" id="id16">
<span id="sangercontig-shiny-dendrogram"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_dendrogram.png"><img alt="../_images/SangerContig_shiny_dendrogram.png" src="../_images/SangerContig_shiny_dendrogram.png" style="width: 660.0px; height: 456.59999999999997px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 7. <em>SangerContig</em> page - dendrogram.</span><a class="headerlink" href="#id16" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p><a class="reference internal" href="#sangercontig-shiny-samples-distance"><span class="std std-ref">Figure 8</span></a> shows distance between <strong>AB1</strong> files. The results are generated by <code class="code docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> function in <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/DECIPHER.html">DECIPHER</a> package. The heatmap is generated by <code class="code docutils literal notranslate"><span class="pre">plot_ly</span></code> function in <a class="reference external" href="https://plot.ly/r/">plotly</a> package.</p>
<figure class="align-center" id="id17">
<span id="sangercontig-shiny-samples-distance"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_samples_distance.png"><img alt="../_images/SangerContig_shiny_samples_distance.png" src="../_images/SangerContig_shiny_samples_distance.png" style="width: 658.8px; height: 455.4px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 8. <em>SangerContig</em> page - samples distance.</span><a class="headerlink" href="#id17" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p><a class="reference internal" href="#sangercontig-shiny-indelsdf-stopcodonsdf"><span class="std std-ref">Figure 9</span></a> shows insertions, deletions and stop codons data frame.</p>
<figure class="align-center" id="id18">
<span id="sangercontig-shiny-indelsdf-stopcodonsdf"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_indelsDF_stopcodonsDF.png"><img alt="../_images/SangerContig_shiny_indelsDF_stopcodonsDF.png" src="../_images/SangerContig_shiny_indelsDF_stopcodonsDF.png" style="width: 631.1999999999999px; height: 436.2px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 9. <em>SangerContig</em> page - indels and stop codons data frame.</span><a class="headerlink" href="#id18" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
<section id="sangerread-page-sc-app">
<h3><em>SangerRead</em> page (SC app)<a class="headerlink" href="#sangerread-page-sc-app" title="Permalink to this headline">¶</a></h3>
<p>Now, let's go to the next level which is also the lowest level, <em>SangerRead</em> page. <em>SangerRead</em> page contains all details of a read including its trimming and chromatogram inputs and results. All reads are in &quot;forward&quot; or &quot;reverse&quot; direction. In this example, there is one read in each direction and <a class="reference internal" href="#sangercontig-shiny-sangerread-page"><span class="std std-ref">Figure 10</span></a> shows &quot;1 Forward Read&quot; page. This page provides basic information, quality trimming inputs, chromatogram plotting inputs etc. Primary/secondary sequences and quality Phred scores table in this figure are dynamic based on the <code class="code docutils literal notranslate"><span class="pre">signalRatioCutoff</span></code> value for base calling and the length of them are always same. Another thing to mention is that primary/secondary sequences and the sequences in the chromatogram in <a class="reference internal" href="#sangercontig-shiny-chromatogram-panel"><span class="std std-ref">Figure 15</span></a> below will always be same after trimming and their color codings for A/T/C/G are same as well.</p>
<figure class="align-center" id="id19">
<span id="sangercontig-shiny-sangerread-page"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_SangerRead_page.png"><img alt="../_images/SangerContig_shiny_SangerRead_page.png" src="../_images/SangerContig_shiny_SangerRead_page.png" style="width: 576.0px; height: 327.20000000000005px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 10. <em>SangerContig</em> Shiny app - <em>SangerRead</em> page</span><a class="headerlink" href="#id19" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>In quality trimming steps, we removes fragment at both ends of sequencing reads with low quality score. It is important because trimmed reads will improves alignment results. <a class="reference internal" href="#sangercontig-shiny-trimming-1"><span class="std std-ref">Figure 11</span></a> shows the UI for Trimming Method 1 (M1): ‘Modified Mott Trimming’. This method is implemented in <a class="reference external" href="http://www.phrap.org/phredphrapconsed.html">Phred</a>. Users can change the cutoff score and click “Apply Trimming Parameters&quot; button to update the UI. The value of input must be between 0 and 1. If the input is invalid, the cutoff score will be set to default 0.0001.</p>
<figure class="align-center" id="id20">
<span id="sangercontig-shiny-trimming-1"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_trimming_1.png"><img alt="../_images/SangerContig_shiny_trimming_1.png" src="../_images/SangerContig_shiny_trimming_1.png" style="width: 693.0px; height: 324.0px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 11. <em>SangerRead</em> page - Trimming Method 1 (M1): ‘Modified Mott Trimming’ UI.</span><a class="headerlink" href="#id20" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p><a class="reference internal" href="#sangercontig-shiny-trimming-2"><span class="std std-ref">Figure 12</span></a> shows another quality trimming method for users to choose from, Trimming Method 2 (M2): ‘Trimmomatics Sliding Window Trimming’. This method is implemented in <a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatics</a>. Users can change the cutoff quality score as well as sliding window size and click “Apply Trimming Parameters&quot; button to update the UI. The value of cutoff quality score must be between 0 and 60 (default 20); the value of sliding window size must be between 0 and 40 (default 10). If the inputs are invalid, their values will be set to default.</p>
<figure class="align-center" id="id21">
<span id="sangercontig-shiny-trimming-2"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_trimming_2.png"><img alt="../_images/SangerContig_shiny_trimming_2.png" src="../_images/SangerContig_shiny_trimming_2.png" style="width: 693.6px; height: 323.4px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 12. <em>SangerRead</em> page - Trimming Method 2 (M2): ‘Trimmomatics Sliding Window Trimming’ UI.</span><a class="headerlink" href="#id21" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p><a class="reference internal" href="#sangercontig-shiny-trimmed-before-after"><span class="std std-ref">Figure 13</span></a> shows the quality report before and after trimming. After clicking the “Apply Trimming Parameters” button in <a class="reference internal" href="#sangercontig-shiny-trimming-1"><span class="std std-ref">Figure 11</span></a> or <a class="reference internal" href="#sangercontig-shiny-trimming-2"><span class="std std-ref">Figure 12</span></a>, the values of these information boxes will be updated to the latest values.</p>
<figure class="align-center" id="id22">
<span id="sangercontig-shiny-trimmed-before-after"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_trimmed_before_after.png"><img alt="../_images/SangerContig_shiny_trimmed_before_after.png" src="../_images/SangerContig_shiny_trimmed_before_after.png" style="width: 660.6px; height: 384.0px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 13. <em>SangerRead</em> page - read quality report before / after trimming.</span><a class="headerlink" href="#id22" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>In <a class="reference internal" href="#sangercontig-shiny-bp-quality-plot"><span class="std std-ref">Figure 14</span></a>, the x-axis is the index of the base pairs; the y-axis is the Phred quality score. The green horizontal bar at the top of the plot is the raw read region and the orange horizontal bar represents the remaining read region. Both <a class="reference internal" href="#sangercontig-shiny-bp-quality-plot"><span class="std std-ref">Figure 14</span></a> trimming plot and <a class="reference internal" href="#sangercontig-shiny-chromatogram-panel"><span class="std std-ref">Figure 15</span></a> chromatogram will be updated once users change the quality trimming parameters and click the “Apply Trimming Parameters&quot; button in <a class="reference internal" href="#sangercontig-shiny-chromatogram-panel"><span class="std std-ref">Figure 15</span></a>.</p>
<figure class="align-center" id="id23">
<span id="sangercontig-shiny-bp-quality-plot"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_bp_quality_plot.png"><img alt="../_images/SangerContig_shiny_bp_quality_plot.png" src="../_images/SangerContig_shiny_bp_quality_plot.png" style="width: 646.1999999999999px; height: 312.0px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 14. <em>SangerContig</em> page - quality trimming plot.</span><a class="headerlink" href="#id23" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>If we only see primary and secondary sequences in the table, we will loose some variations. Chromatogram is very helpful to check the peak resolution. <a class="reference internal" href="#sangercontig-shiny-chromatogram-panel"><span class="std std-ref">Figure 15</span></a> shows the panel of plotting chromatogram. Users can change four parameters: <code class="code docutils literal notranslate"><span class="pre">Base</span> <span class="pre">Number</span> <span class="pre">Per</span> <span class="pre">Row</span></code>, <code class="code docutils literal notranslate"><span class="pre">Height</span> <span class="pre">Per</span> <span class="pre">Row</span></code>, <code class="code docutils literal notranslate"><span class="pre">Signal</span> <span class="pre">Ratio</span> <span class="pre">Cutoff</span></code>, and <code class="code docutils literal notranslate"><span class="pre">Show</span> <span class="pre">Trimmed</span> <span class="pre">Region</span></code>. Among them, <code class="code docutils literal notranslate"><span class="pre">Signal</span> <span class="pre">Ratio</span> <span class="pre">Cutoff</span></code> is a key parameter. If its value is default value 0.33, it indicates that the lower peak should be at least 1/3rd as high as the higher peak for it count as a secondary peak.</p>
<figure class="align-center" id="id24">
<span id="sangercontig-shiny-chromatogram-panel"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_chromatogram_panel.png"><img alt="../_images/SangerContig_shiny_chromatogram_panel.png" src="../_images/SangerContig_shiny_chromatogram_panel.png" style="width: 693.6px; height: 488.4px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 15. <em>SangerContig</em> page - chromatogram panel.</span><a class="headerlink" href="#id24" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>Here is an example of applying new chromatogram parameters. We click “Show Trimmed Region” to set its value from <code class="code docutils literal notranslate"><span class="pre">FALSE</span></code> to <code class="code docutils literal notranslate"><span class="pre">TRUE</span></code> and click the &quot;Apply Chromatogram Parameters&quot; button. <a class="reference internal" href="#sangercontig-plotting-popup"><span class="std std-ref">Figure 16</span></a> shows the loading notification popup during base calling and chromatogram plotting.</p>
<figure class="align-center" id="id25">
<span id="sangercontig-plotting-popup"></span><a class="reference internal image-reference" href="../_images/SangerContig_plotting_popup.png"><img alt="../_images/SangerContig_plotting_popup.png" src="../_images/SangerContig_plotting_popup.png" style="width: 691.1999999999999px; height: 487.79999999999995px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 16. <em>SangerContig</em> page - loading notification popup during replotting chromatogram.</span><a class="headerlink" href="#id25" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>After replotting the chromatogram, we can see that trimmed region is showed in red striped region. <a class="reference internal" href="#sangercontig-shiny-chromatogram"><span class="std std-ref">Figure 17</span></a> shows part of the the chromatogram (1 bp ~ 240 bp). Moreover, chromatogram will be replotted when trimmed positions or chromatogram parameters are updated.</p>
<figure class="align-center" id="id26">
<span id="sangercontig-shiny-chromatogram"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_chromatogram.png"><img alt="../_images/SangerContig_shiny_chromatogram.png" src="../_images/SangerContig_shiny_chromatogram.png" style="width: 660.0px; height: 427.8px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 17. <em>SangerContig</em> page - chromatogram with trimmed region showed.</span><a class="headerlink" href="#id26" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>To let users browse the trimmed primary/secondary sequences without finding “Trimming Start Point” and “Trimming End Point” by themselves, we provide the final trimmed primary/secondary sequences that will be used for reads alignment with quality scores in table format in <a class="reference internal" href="#sangercontig-shiny-trimmed-sequences"><span class="std std-ref">Figure 18</span></a>. Frameshift amino acid sequences are also provided.</p>
<figure class="align-center" id="id27">
<span id="sangercontig-shiny-trimmed-sequences"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_trimmed_sequences.png"><img alt="../_images/SangerContig_shiny_trimmed_sequences.png" src="../_images/SangerContig_shiny_trimmed_sequences.png" style="width: 646.8000000000001px; height: 298.48px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 18. <em>SangerContig</em> page - trimmed primary/secondary sequences and Phred quality score in table format.</span><a class="headerlink" href="#id27" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>We have updated the trimming and chromatogram parameters for each read. Now, we need to click “Re-calculate contig” button to do alignment again. Last but not least, we can save all data into a new ‘SangerContig’ S4 instance by clicking “Save S4 Instance button”. New S4 instance will be saved in <strong>Rda</strong> format. Users can run <code class="code docutils literal notranslate"><span class="pre">readRDS</span></code> function to load it into current R environment. <a class="reference internal" href="#sangercontig-shiny-save-popup"><span class="std std-ref">Figure 19</span></a> shows some hints in the save notification popup.</p>
<figure class="align-center" id="id28">
<span id="sangercontig-shiny-save-popup"></span><a class="reference internal image-reference" href="../_images/SangerContig_shiny_save_popup.png"><img alt="../_images/SangerContig_shiny_save_popup.png" src="../_images/SangerContig_shiny_save_popup.png" style="width: 719.0px; height: 437.0px;" /></a>
<figcaption>
<p><span class="caption-text">Figure 19. <em>SangerContig</em> page - saving notification popup.</span><a class="headerlink" href="#id28" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="writing-sangercontig-fasta-files-ab1">
<h2>Writing <em>SangerContig</em> FASTA files <sub>(AB1)</sub><a class="headerlink" href="#writing-sangercontig-fasta-files-ab1" title="Permalink to this headline">¶</a></h2>
<p>Users can write the <em>SangerContig</em> instance, <code class="code docutils literal notranslate"><span class="pre">my_sangerContig</span></code>, to <strong>FASTA</strong> files. There are four options for users to choose from in <code class="code docutils literal notranslate"><span class="pre">selection</span></code> parameter.</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">reads_unalignment</span></code>: Writing reads into a single <strong>FASTA</strong> file (only trimmed without alignment).</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">reads_alignment</span></code>: Writing reads alignment and contig read to a single <strong>FASTA</strong> file.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">contig</span></code>: Writing the contig to a single <strong>FASTA</strong> file.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">all</span></code>: Writing reads, reads alignment, and the contig into three different files.</p></li>
</ul>
<p>Below is the oneliner for writing out <strong>FASTA</strong> files. This function mainly depends on <code class="code docutils literal notranslate"><span class="pre">writeXStringSet</span></code> function in <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/Biostrings.html">Biostrings</a> R package. Users can set the compression level through <code class="code docutils literal notranslate"><span class="pre">writeFasta</span></code> function.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">writeFasta</span><span class="p">(</span><span class="n">my_sangerContig</span><span class="p">,</span>
           <span class="n">outputDir</span>         <span class="o">=</span> <span class="nf">tempdir</span><span class="p">(),</span>
           <span class="n">compress</span>          <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>
           <span class="n">compression_level</span> <span class="o">=</span> <span class="kc">NA</span><span class="p">,</span>
           <span class="n">selection</span>         <span class="o">=</span> <span class="s">&quot;all&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Users can download the output FASTA file of this example through the following three links:</p>
<ol class="arabic simple">
<li><p><a class="reference download internal" download="" href="../_downloads/95ef9378c1407481d69c3bf882a580bb/Achl_RBNII384-13_reads_unalignment.fa"><code class="xref download docutils literal notranslate"><span class="pre">Achl_RBNII384-13_reads_unalignment.fa</span></code></a></p></li>
<li><p><a class="reference download internal" download="" href="../_downloads/eeb554b575ee5b29ef733dafbf876c36/Achl_RBNII384-13_reads_alignment.fa"><code class="xref download docutils literal notranslate"><span class="pre">Achl_RBNII384-13_reads_alignment.fa</span></code></a></p></li>
<li><p><a class="reference download internal" download="" href="../_downloads/345ac400befc4f176c75732770ec3fcb/Achl_RBNII384-13_contig.fa"><code class="xref download docutils literal notranslate"><span class="pre">Achl_RBNII384-13_contig.fa</span></code></a></p></li>
</ol>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="generating-sangercontig-report-ab1">
<h2>Generating <em>SangerContig</em> report <sub>(AB1)</sub><a class="headerlink" href="#generating-sangercontig-report-ab1" title="Permalink to this headline">¶</a></h2>
<p>Last but not least, users can save <em>SangerContig</em> instance, <code class="code docutils literal notranslate"><span class="pre">my_sangerContig</span></code>, into a report after the analysis. The report will be generated in <strong>HTML</strong> by knitting <strong>Rmd</strong> files.</p>
<p>Users can set <code class="code docutils literal notranslate"><span class="pre">includeSangerRead</span></code> parameter to decide to which level the <em>SangerContig</em> report will go. Moreover, after the reports are generated,
users can easily navigate through reports in different levels within the <strong>HTML</strong> file.</p>
<p>One thing to pay attention to is that if users have many reads, it will take quite a long time to write out all reports. If users only want to generate the contig result, remember to set <code class="code docutils literal notranslate"><span class="pre">includeSangerRead</span></code> to <code class="code docutils literal notranslate"><span class="pre">FALSE</span></code> in order to save time.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">generateReport</span><span class="p">(</span><span class="n">my_sangerContig</span><span class="p">,</span>
               <span class="n">outputDir</span>           <span class="o">=</span> <span class="nf">tempdir</span><span class="p">(),</span>
               <span class="n">includeSangerRead</span>   <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<p>Here is the generated <a class="reference external" href="https://kuanhao-chao.github.io/sangeranalyseR_report/SangerContig/AB1/Achl_RBNII384-13/SangerContig_Report.html">SangerContig html report of this example (ABIF)</a>. Users can access to '<em>Basic Information</em>', '<em>SangerContig Input Parameters</em>', '<em>Contig Sequence</em>' and '<em>Contig Results</em>' sections inside it. Furthermore, users can also navigate through html reports of all forward and reverse <em>SangerRead</em> in this <em>SangerContig</em> report.</p>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
</section>
<section id="code-summary-sangercontig-ab1">
<h2>Code summary (<em>SangerContig</em>, <strong>AB1</strong>)<a class="headerlink" href="#code-summary-sangercontig-ab1" title="Permalink to this headline">¶</a></h2>
<section id="id3">
<h3>(1) Preparing <em>SangerContig</em> <strong>AB1</strong> inputs<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">rawDataDir</span> <span class="o">&lt;-</span> <span class="nf">system.file</span><span class="p">(</span><span class="s">&quot;extdata&quot;</span><span class="p">,</span> <span class="n">package</span> <span class="o">=</span> <span class="s">&quot;sangeranalyseR&quot;</span><span class="p">)</span>
<span class="n">parentDir</span> <span class="o">&lt;-</span> <span class="nf">file.path</span><span class="p">(</span><span class="n">rawDataDir</span><span class="p">,</span> <span class="s">&quot;Allolobophora_chlorotica&quot;</span><span class="p">,</span> <span class="s">&quot;RBNII&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="id4">
<h3>(2) Creating <em>SangerContig</em> instance from <strong>AB1</strong><a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<section id="regular-expression-method-sangercontig-creation-ab1">
<h4>(2.1) &quot;Regular Expression Method&quot; <em>SangerContig</em> creation (<strong>AB1</strong>)<a class="headerlink" href="#regular-expression-method-sangercontig-creation-ab1" title="Permalink to this headline">¶</a></h4>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># using `constructor` function to create SangerContig instance</span>
<span class="n">my_sangerContig</span> <span class="o">&lt;-</span> <span class="nf">SangerContig</span><span class="p">(</span><span class="n">inputSource</span>           <span class="o">=</span> <span class="s">&quot;ABIF&quot;</span><span class="p">,</span>
                                <span class="n">processMethod</span>         <span class="o">=</span> <span class="s">&quot;REGEX&quot;</span><span class="p">,</span>
                                <span class="n">ABIF_Directory</span>        <span class="o">=</span> <span class="n">parentDir</span><span class="p">,</span>
                                <span class="n">contigName</span>            <span class="o">=</span> <span class="s">&quot;Achl_RBNII384-13&quot;</span><span class="p">,</span>
                                <span class="n">REGEX_SuffixForward</span>   <span class="o">=</span> <span class="s">&quot;_[0-9]*_F.ab1$&quot;</span><span class="p">,</span>
                                <span class="n">REGEX_SuffixReverse</span>   <span class="o">=</span> <span class="s">&quot;_[0-9]*_R.ab1$&quot;</span><span class="p">,</span>
                                <span class="n">refAminoAcidSeq</span> <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">)</span>

<span class="c1"># using `new` method to create SangerContig instance</span>
<span class="n">my_sangerContig</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;SangerContig&quot;</span><span class="p">,</span>
                       <span class="n">inputSource</span>           <span class="o">=</span> <span class="s">&quot;ABIF&quot;</span><span class="p">,</span>
                       <span class="n">processMethod</span>         <span class="o">=</span> <span class="s">&quot;REGEX&quot;</span><span class="p">,</span>
                       <span class="n">ABIF_Directory</span>        <span class="o">=</span> <span class="n">parentDir</span><span class="p">,</span>
                       <span class="n">contigName</span>            <span class="o">=</span> <span class="s">&quot;Achl_RBNII384-13&quot;</span><span class="p">,</span>
                       <span class="n">REGEX_SuffixForward</span>   <span class="o">=</span> <span class="s">&quot;_[0-9]*_F.ab1$&quot;</span><span class="p">,</span>
                       <span class="n">REGEX_SuffixReverse</span>   <span class="o">=</span> <span class="s">&quot;_[0-9]*_R.ab1$&quot;</span><span class="p">,</span>
                       <span class="n">refAminoAcidSeq</span> <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="toggle docutils container">
<div class="header docutils container">
<p>Following is the R shell output that you will get.</p>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>INFO [2021-29-06 17:44:10] ========================================================
INFO [2021-29-06 17:44:10] ================ Creating &#39;SangerContig&#39; ===============
INFO [2021-29-06 17:44:10] ========================================================
INFO [2021-29-06 17:44:10]   &gt;&gt; Contig Name: &#39;Achl_RBNII384-13&#39;
INFO [2021-29-06 17:44:10]   &gt;&gt; You are using Regular Expression Method to group AB1 files!
INFO [2021-29-06 17:44:10] &gt;&gt; Your contig name is Achl_RBNII384-13
SUCCESS [2021-29-06 17:44:11] --------------------------------------------------------
SUCCESS [2021-29-06 17:44:11] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-29-06 17:44:11] --------------------------------------------------------
SUCCESS [2021-29-06 17:44:11]    &gt;&gt; &#39;Achl_RBNII384-13_1_F.ab1&#39; is created (Forward Read; ABIF).
SUCCESS [2021-29-06 17:44:12] --------------------------------------------------------
SUCCESS [2021-29-06 17:44:12] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-29-06 17:44:12] --------------------------------------------------------
SUCCESS [2021-29-06 17:44:12]    &gt;&gt; &#39;Achl_RBNII384-13_2_R.ab1&#39; is created (Reverse Read; ABIF).
INFO [2021-29-06 17:44:12]    &gt;&gt; The number of reads detected: 2
INFO [2021-29-06 17:44:12] Correcting frameshifts in reads using amino acidreference sequence
Assessing frameshifts in nucleotide sequences:
|=============================================================================| 100%

Time difference of 0.2 secs
SUCCESS [2021-29-06 17:44:16] ==========================================================
SUCCESS [2021-29-06 17:44:16] ======== &#39;SangerContig&#39; S4 instance is created !! ========
SUCCESS [2021-29-06 17:44:16] ==========================================================
INFO [2021-29-06 17:44:16]    &gt;&gt; 2 read(s) created from ABIF file.
INFO [2021-29-06 17:44:16]      &gt;&gt; 1 reads assigned to &#39;forward reads&#39; according to &#39;regular expression&#39;.
INFO [2021-29-06 17:44:16]      &gt;&gt; 1 reads assigned to &#39;reverse reads&#39; according to &#39;regular expression&#39;.
INFO [2021-29-06 17:44:16]    &gt;&gt; Trimmed by &#39;M1 - Mott’s trimming algorithm&#39;.
DEBUG [2021-29-06 17:44:16]    &gt;&gt; For more information, please run &#39;object&#39;
DEBUG [2021-29-06 17:44:16]    &gt;&gt; Run &#39;object@objectResults@readResultTable&#39; to check the results of each Sanger reads
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="id5">
<h4>(2.2) &quot;CSV file matching&quot; <em>SangerContig</em> creation (<strong>AB1</strong>)<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h4>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">csv_namesConversion</span> <span class="o">&lt;-</span> <span class="nf">file.path</span><span class="p">(</span><span class="n">rawDataDir</span><span class="p">,</span> <span class="s">&quot;ab1&quot;</span><span class="p">,</span> <span class="s">&quot;SangerContig&quot;</span><span class="p">,</span> <span class="s">&quot;names_conversion_2.csv&quot;</span><span class="p">)</span>

<span class="c1"># using `constructor` function to create SangerContig instance</span>
<span class="n">my_sangerContig</span> <span class="o">&lt;-</span> <span class="nf">SangerContig</span><span class="p">(</span><span class="n">inputSource</span>           <span class="o">=</span> <span class="s">&quot;ABIF&quot;</span><span class="p">,</span>
                                <span class="n">processMethod</span>         <span class="o">=</span> <span class="s">&quot;CSV&quot;</span><span class="p">,</span>
                                <span class="n">ABIF_Directory</span>        <span class="o">=</span> <span class="n">parentDir</span><span class="p">,</span>
                                <span class="n">contigName</span>            <span class="o">=</span> <span class="s">&quot;Achl_RBNII384-13&quot;</span><span class="p">,</span>
                                <span class="n">CSV_NamesConversion</span>   <span class="o">=</span> <span class="n">csv_namesConversion</span><span class="p">,</span>
                                <span class="n">refAminoAcidSeq</span> <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">)</span>


<span class="c1"># using `new` method to create SangerContig instance</span>
<span class="n">my_sangerContig</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;SangerContig&quot;</span><span class="p">,</span>
                       <span class="n">inputSource</span>           <span class="o">=</span> <span class="s">&quot;ABIF&quot;</span><span class="p">,</span>
                       <span class="n">processMethod</span>         <span class="o">=</span> <span class="s">&quot;CSV&quot;</span><span class="p">,</span>
                       <span class="n">ABIF_Directory</span>        <span class="o">=</span> <span class="n">parentDir</span><span class="p">,</span>
                       <span class="n">contigName</span>            <span class="o">=</span> <span class="s">&quot;Achl_RBNII384-13&quot;</span><span class="p">,</span>
                       <span class="n">CSV_NamesConversion</span>   <span class="o">=</span> <span class="n">csv_namesConversion</span><span class="p">,</span>
                       <span class="n">refAminoAcidSeq</span> <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="toggle docutils container">
<div class="header docutils container">
<p>Following is the R shell output that you will get.</p>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>WARN [2021-09-07 01:47:45] &#39;Achl_RBNII395-13_1_F.ab1&#39; is not in the csv file (/Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata/ab1/SangerContig/names_conversion_2.csv)
WARN [2021-09-07 01:47:45] &#39;Achl_RBNII395-13_2_R.ab1&#39; is not in the csv file (/Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata/ab1/SangerContig/names_conversion_2.csv)
WARN [2021-09-07 01:47:45] &#39;Achl_RBNII396-13_1_F.ab1&#39; is not in the csv file (/Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata/ab1/SangerContig/names_conversion_2.csv)
WARN [2021-09-07 01:47:45] &#39;Achl_RBNII396-13_2_R.ab1&#39; is not in the csv file (/Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata/ab1/SangerContig/names_conversion_2.csv)
WARN [2021-09-07 01:47:45] &#39;Achl_RBNII397-13_1_F.ab1&#39; is not in the csv file (/Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata/ab1/SangerContig/names_conversion_2.csv)
WARN [2021-09-07 01:47:45] &#39;Achl_RBNII397-13_2_R.ab1&#39; is not in the csv file (/Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata/ab1/SangerContig/names_conversion_2.csv)
INFO [2021-09-07 01:47:45] ========================================================
INFO [2021-09-07 01:47:45] ================ Creating &#39;SangerContig&#39; ===============
INFO [2021-09-07 01:47:45] ========================================================
INFO [2021-09-07 01:47:45]   &gt;&gt; Contig Name: &#39;Achl_RBNII384-13&#39;
INFO [2021-09-07 01:47:45]   &gt;&gt; You are using CSV Name Conversion Method to group AB1 files!
INFO [2021-09-07 01:47:45] &gt;&gt; Your contig name is Achl_RBNII384-13
SUCCESS [2021-09-07 01:47:46] --------------------------------------------------------
SUCCESS [2021-09-07 01:47:46] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-09-07 01:47:46] --------------------------------------------------------
SUCCESS [2021-09-07 01:47:46]    &gt;&gt; &#39;Achl_RBNII384-13_1_F.ab1&#39; is created (Forward Read; ABIF).
SUCCESS [2021-09-07 01:47:47] --------------------------------------------------------
SUCCESS [2021-09-07 01:47:47] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-09-07 01:47:47] --------------------------------------------------------
SUCCESS [2021-09-07 01:47:47]    &gt;&gt; &#39;Achl_RBNII384-13_2_R.ab1&#39; is created (Reverse Read; ABIF).
INFO [2021-09-07 01:47:47]    &gt;&gt; The number of reads detected: 2
INFO [2021-09-07 01:47:47] Correcting frameshifts in reads using amino acidreference sequence
Assessing frameshifts in nucleotide sequences:
  |================================================================================================| 100%

Time difference of 0.08 secs
SUCCESS [2021-09-07 01:47:51] ==========================================================
SUCCESS [2021-09-07 01:47:51] ======== &#39;SangerContig&#39; S4 instance is created !! ========
SUCCESS [2021-09-07 01:47:51] ==========================================================
INFO [2021-09-07 01:47:51]    &gt;&gt; 2 read(s) created from ABIF file.
INFO [2021-09-07 01:47:51]      &gt;&gt; 1 reads assigned to &#39;forward reads&#39; according to &#39;csv file&#39;.
INFO [2021-09-07 01:47:51]      &gt;&gt; 1 reads assigned to &#39;reverse reads&#39; according to &#39;csv file&#39;.
INFO [2021-09-07 01:47:51]    &gt;&gt; Trimmed by &#39;M1 - Mott’s trimming algorithm&#39;.
DEBUG [2021-09-07 01:47:51]    &gt;&gt; For more information, please run &#39;object&#39;
DEBUG [2021-09-07 01:47:51]    &gt;&gt; Run &#39;object@objectResults@readResultTable&#39; to check the results of each Sanger reads
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="id6">
<h3>(3) Updating <em>SangerContig</em> quality trimming parameters<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">newSangerContig</span> <span class="o">&lt;-</span> <span class="nf">updateQualityParam</span><span class="p">(</span><span class="n">my_sangerContig</span><span class="p">,</span>
                                      <span class="n">TrimmingMethod</span>       <span class="o">=</span> <span class="s">&quot;M2&quot;</span><span class="p">,</span>
                                      <span class="n">M1TrimmingCutoff</span>     <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                                      <span class="n">M2CutoffQualityScore</span> <span class="o">=</span> <span class="m">20</span><span class="p">,</span>
                                      <span class="n">M2SlidingWindowSize</span>  <span class="o">=</span> <span class="m">15</span><span class="p">)</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="id7">
<h3>(4) Launching <em>SangerContig</em> Shiny app<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">launchApp</span><span class="p">(</span><span class="n">my_sangerContig</span><span class="p">)</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="id8">
<h3>(5) Writing <em>SangerContig</em> FASTA files <sub>(AB1)</sub><a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">writeFasta</span><span class="p">(</span><span class="n">my_sangerContig</span><span class="p">)</span>
</pre></div>
</div>
<div class="toggle docutils container">
<div class="header docutils container">
<p>Following is the R shell output that you will get.</p>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">29</span><span class="o">-</span><span class="mi">06</span> <span class="mi">17</span><span class="p">:</span><span class="mi">47</span><span class="p">:</span><span class="mi">19</span><span class="p">]</span> <span class="n">Your</span> <span class="nb">input</span> <span class="ow">is</span> <span class="s1">&#39;SangerContig&#39;</span> <span class="n">S4</span> <span class="n">instance</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">29</span><span class="o">-</span><span class="mi">06</span> <span class="mi">17</span><span class="p">:</span><span class="mi">47</span><span class="p">:</span><span class="mi">19</span><span class="p">]</span> <span class="o">&gt;&gt;&gt;</span> <span class="n">outputDir</span> <span class="p">:</span> <span class="o">/</span><span class="n">private</span><span class="o">/</span><span class="n">var</span><span class="o">/</span><span class="n">folders</span><span class="o">/</span><span class="mi">33</span><span class="o">/</span><span class="mi">7</span><span class="n">v38jdjd2874jcxb6l71m00h0000gn</span><span class="o">/</span><span class="n">T</span><span class="o">/</span><span class="n">RtmpRAPaMV</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">29</span><span class="o">-</span><span class="mi">06</span> <span class="mi">17</span><span class="p">:</span><span class="mi">47</span><span class="p">:</span><span class="mi">19</span><span class="p">]</span> <span class="n">Start</span> <span class="n">to</span> <span class="n">write</span> <span class="s1">&#39;Achl_RBNII384-13&#39;</span> <span class="n">to</span> <span class="n">FASTA</span> <span class="nb">format</span> <span class="o">...</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">29</span><span class="o">-</span><span class="mi">06</span> <span class="mi">17</span><span class="p">:</span><span class="mi">47</span><span class="p">:</span><span class="mi">19</span><span class="p">]</span> <span class="o">&gt;&gt;</span> <span class="n">Writing</span> <span class="n">alignment</span> <span class="n">to</span> <span class="n">FASTA</span> <span class="o">...</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">29</span><span class="o">-</span><span class="mi">06</span> <span class="mi">17</span><span class="p">:</span><span class="mi">47</span><span class="p">:</span><span class="mi">19</span><span class="p">]</span> <span class="o">&gt;&gt;</span> <span class="n">Writing</span> <span class="nb">all</span> <span class="n">single</span> <span class="n">reads</span> <span class="n">to</span> <span class="n">FASTA</span> <span class="o">...</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">29</span><span class="o">-</span><span class="mi">06</span> <span class="mi">17</span><span class="p">:</span><span class="mi">47</span><span class="p">:</span><span class="mi">19</span><span class="p">]</span> <span class="o">&gt;&gt;</span> <span class="n">Writing</span> <span class="n">consensus</span> <span class="n">read</span> <span class="n">to</span> <span class="n">FASTA</span> <span class="o">...</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">29</span><span class="o">-</span><span class="mi">06</span> <span class="mi">17</span><span class="p">:</span><span class="mi">47</span><span class="p">:</span><span class="mi">19</span><span class="p">]</span> <span class="n">Finish</span> <span class="n">writing</span> <span class="s1">&#39;Achl_RBNII384-13&#39;</span> <span class="n">to</span> <span class="n">FASTA</span> <span class="nb">format</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<p>You will get three FASTA files:</p>
<ol class="arabic simple">
<li><p><a class="reference download internal" download="" href="../_downloads/95ef9378c1407481d69c3bf882a580bb/Achl_RBNII384-13_reads_unalignment.fa"><code class="xref download docutils literal notranslate"><span class="pre">Achl_RBNII384-13_reads_unalignment.fa</span></code></a></p></li>
<li><p><a class="reference download internal" download="" href="../_downloads/eeb554b575ee5b29ef733dafbf876c36/Achl_RBNII384-13_reads_alignment.fa"><code class="xref download docutils literal notranslate"><span class="pre">Achl_RBNII384-13_reads_alignment.fa</span></code></a></p></li>
<li><p><a class="reference download internal" download="" href="../_downloads/345ac400befc4f176c75732770ec3fcb/Achl_RBNII384-13_contig.fa"><code class="xref download docutils literal notranslate"><span class="pre">Achl_RBNII384-13_contig.fa</span></code></a></p></li>
</ol>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="id9">
<h3>(6) Generating <em>SangerContig</em> report <sub>(AB1)</sub><a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">generateReport</span><span class="p">(</span><span class="n">my_sangerContig</span><span class="p">)</span>
</pre></div>
</div>
<p>You can check the html report of <a class="reference external" href="https://kuanhao-chao.github.io/sangeranalyseR_report/SangerContig/AB1/Achl_RBNII384-13/SangerContig_Report.html">this SangerContig example (ABIF)</a>.</p>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
</section>
</section>
</section>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="advance_user_SangerAlignment_ab1.html" class="btn btn-neutral float-right" title="Advanced User Guide - SangerAlignment (AB1)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="advance_user_SangerRead_ab1.html" class="btn btn-neutral float-left" title="Advanced User Guide - SangerRead (AB1)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Kuan-Hao Chao

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 <script type="text/javascript">
    $(document).ready(function() {
        $(".toggle > *").hide();
        $(".toggle .header").show();
        $(".toggle .header").click(function() {
            $(this).parent().children().not(".header").toggle(400);
            $(this).parent().children(".header").toggleClass("open");
        })
    });
</script>


</body>
</html>