

<!DOCTYPE html>
<html class="writer-html5" lang="English" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Advanced User Guide - SangerAlignment (FASTA) &mdash; sangeranalyseR  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Q &amp; A ..." href="how_to_page.html" />
    <link rel="prev" title="Advanced User Guide - SangerContig (FASTA)" href="advance_user_SangerContig_fasta.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> sangeranalyseR
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Beginners Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerRead_ab1.html">Advanced User Guide - <em>SangerRead</em> (<strong>AB1</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerContig_ab1.html">Advanced User Guide - <em>SangerContig</em> (<strong>AB1</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerAlignment_ab1.html">Advanced User Guide - <em>SangerAlignment</em> (<strong>AB1</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerRead_fasta.html">Advanced User Guide - <em>SangerRead</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerContig_fasta.html">Advanced User Guide - <em>SangerContig</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Advanced User Guide - <em>SangerAlignment</em> (<strong>FASTA</strong>)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#preparing-sangeralignment-fasta-input">Preparing <em>SangerAlignment</em> <strong>FASTA</strong> input</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#regular-expression-matching-sangeralignment-inputs-fasta">(1) &quot;regular expression matching&quot; <em>SangerAlignment</em> inputs (<strong>FASTA</strong>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#csv-file-matching-sangeralignment-inputs-fasta">(2) &quot;CSV file matching&quot; <em>SangerAlignment</em> inputs (<strong>FASTA</strong>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#creating-sangeralignment-instance-from-fasta">Creating <em>SangerAlignment</em> instance from <strong>FASTA</strong></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#regular-expression-matching-sangeralignment-creation-fasta">(1) &quot;regular expression matching&quot; <em>SangerAlignment</em> creation (<strong>FASTA</strong>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#csv-file-matching-sangeralignment-creation-fasta">(2) &quot;CSV file matching&quot; <em>SangerAlignment</em> creation (<strong>FASTA</strong>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#writing-sangeralignment-fasta-files-fasta">Writing <em>SangerAlignment</em> FASTA files <sub>(FASTA)</sub></a></li>
<li class="toctree-l2"><a class="reference internal" href="#generating-sangeralignment-report-fasta">Generating <em>SangerAlignment</em> report <sub>(FASTA)</sub></a></li>
<li class="toctree-l2"><a class="reference internal" href="#code-summary-sangeralignment-fasta">Code summary (<em>SangerAlignment</em>, <strong>FASTA</strong>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#preparing-sangeralignment-fasta-inputs">(1) Preparing <em>SangerAlignment</em> <strong>FASTA</strong> inputs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id1">(2) Creating <em>SangerAlignment</em> instance from <strong>FASTA</strong></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#regular-expression-method-sangeralignment-creation-fasta">(2.1) &quot;Regular Expression Method&quot; <em>SangerAlignment</em> creation (<strong>FASTA</strong>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id2">(2.2) &quot;CSV file matching&quot; <em>SangerAlignment</em> creation (<strong>FASTA</strong>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id3">(3) Writing <em>SangerAlignment</em> FASTA files <sub>(FASTA)</sub></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">(4) Generating <em>SangerAlignment</em> report <sub>(FASTA)</sub></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="how_to_page.html">Q &amp; A ...</a></li>
<li class="toctree-l1"><a class="reference internal" href="function_manual.html">User Manual (functions)</a></li>
<li class="toctree-l1"><a class="reference internal" href="frequent_questions.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="conclusion.html">Conclusion</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="contact.html">Contact</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Help</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">sangeranalyseR</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Advanced User Guide - <em>SangerAlignment</em> (<strong>FASTA</strong>)</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/content/advance_user_SangerAlignment_fasta.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="advanced-user-guide-sangeralignment-fasta">
<h1>Advanced User Guide - <em>SangerAlignment</em> (<strong>FASTA</strong>)<a class="headerlink" href="#advanced-user-guide-sangeralignment-fasta" title="Permalink to this headline">¶</a></h1>
<p><em>SangerAlignment</em> is in the toppest level of sangeranalyseR (<a class="reference internal" href="#sangeralignment-hierachy-fasta"><span class="std std-ref">Figure_1</span></a>), and each <strong>SangerAlignment</strong> instance corresponds to an alignment of contigs in a Sanger sequencing experiment. Among its slots, there is a <em>SangerContig</em> list which will be aligned into a consensus contig. Users can access to each <em>SangerContig</em> and <em>SangerRead</em> inside a <em>SangerAlignment</em> instance.</p>
<p>In this section, we are going to go through details about a reproducible <em>SangerAlignment</em> analysis example with the <strong>FASTA</strong> file input in sangeranalyseR. By running the following example codes, you will get an end-to-end <em>SangerAlignment</em> analysis result.</p>
<div class="figure align-center" id="id5">
<span id="sangeralignment-hierachy-fasta"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_hierachy.png"><img alt="../_images/SangerAlignment_hierachy.png" src="../_images/SangerAlignment_hierachy.png" style="width: 545.4px; height: 218.8px;" /></a>
<p class="caption"><span class="caption-text">Figure 1. Classes hierarchy in sangeranalyseR, <em>SangerAlignment</em> level.</span><a class="headerlink" href="#id5" title="Permalink to this image">¶</a></p>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<div class="section" id="preparing-sangeralignment-fasta-input">
<h2>Preparing <em>SangerAlignment</em> <strong>FASTA</strong> input<a class="headerlink" href="#preparing-sangeralignment-fasta-input" title="Permalink to this headline">¶</a></h2>
<p>In <a class="reference internal" href="advance_user_SangerAlignment_ab1.html#advanced-user-guide-sangeralignment-ab1"><span class="std std-ref">Advanced User Guide - SangerAlignment (AB1)</span></a>, we demonstrated how to use <strong>AB1</strong> input files to create <em>SangerAlignment</em> instance. Here, we explain another input format - the <strong>FASTA</strong> input. Before starting the analysis, users need to prepare one <strong>FASTA</strong> file, which must end with <strong>.fa</strong> or <strong>.fasta</strong>, containing sequences of all reads. In this example, the <strong>FASTA</strong> file is in the sangeranalyseR package, and you can simply get its path by running the following codes:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">rawDataDir</span> <span class="o">&lt;-</span> <span class="nf">system.file</span><span class="p">(</span><span class="s">&quot;extdata&quot;</span><span class="p">,</span> <span class="n">package</span> <span class="o">=</span> <span class="s">&quot;sangeranalyseR&quot;</span><span class="p">)</span>
<span class="n">fastaFN</span> <span class="o">&lt;-</span> <span class="nf">file.path</span><span class="p">(</span><span class="n">rawDataDir</span><span class="p">,</span> <span class="s">&quot;fasta&quot;</span><span class="p">,</span> <span class="s">&quot;SangerAlignment&quot;</span><span class="p">,</span> <span class="s">&quot;Sanger_all_reads.fa&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The value of <code class="code docutils literal notranslate"><span class="pre">fastaFN</span></code> is where the <strong>FASTA</strong> file is placed. If your operating system is macOS, then its value should look like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">Library</span><span class="o">/</span><span class="n">Frameworks</span><span class="o">/</span><span class="n">R</span><span class="o">.</span><span class="n">framework</span><span class="o">/</span><span class="n">Versions</span><span class="o">/</span><span class="mf">4.0</span><span class="o">/</span><span class="n">Resources</span><span class="o">/</span><span class="n">library</span><span class="o">/</span><span class="n">sangeranalyseR</span><span class="o">/</span><span class="n">extdata</span><span class="o">/</span><span class="n">fasta</span><span class="o">/</span><span class="n">SangerAlignment</span><span class="o">/</span><span class="n">Sanger_all_reads</span><span class="o">.</span><span class="n">fa</span>
</pre></div>
</div>
<p>And we showed the reads in <code class="code docutils literal notranslate"><span class="pre">fastaFN</span></code> in Figure_2 (<a class="reference download internal" download="" href="../_downloads/bfd858bb8d73620d1300d2e3b2a34899/Sanger_all_reads.fa"><code class="xref download docutils literal notranslate"><span class="pre">example</span> <span class="pre">FASTA</span> <span class="pre">file</span></code></a>):</p>
<div class="figure align-center" id="id6">
<span id="sangeralignment-fasta-input"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_fasta_input.png"><img alt="../_images/SangerAlignment_fasta_input.png" src="../_images/SangerAlignment_fasta_input.png" style="width: 612.9px; height: 641.7px;" /></a>
<p class="caption"><span class="caption-text">Figure 2. <em>SangerAlignment</em> <strong>FASTA</strong> input file (4 out of 8 reads are showed).</span><a class="headerlink" href="#id6" title="Permalink to this image">¶</a></p>
</div>
<p>Inside the <strong>FASTA</strong> file (<a class="reference internal" href="#sangeralignment-fasta-input"><span class="std std-ref">Figure_2</span></a>; <a class="reference download internal" download="" href="../_downloads/bfd858bb8d73620d1300d2e3b2a34899/Sanger_all_reads.fa"><code class="xref download docutils literal notranslate"><span class="pre">Sanger_all_reads.fa</span></code></a>), the strings starting with &quot;&gt;&quot; before each read are the read names. There are two ways of grouping reads which are <strong>&quot;regular expression matching&quot;</strong> and <strong>&quot;CSV file matching&quot;</strong>, and following are instructions of how to prepare your <strong>FASTA</strong> input file.</p>
<div class="section" id="regular-expression-matching-sangeralignment-inputs-fasta">
<h3>(1) &quot;regular expression matching&quot; <em>SangerAlignment</em> inputs (<strong>FASTA</strong>)<a class="headerlink" href="#regular-expression-matching-sangeralignment-inputs-fasta" title="Permalink to this headline">¶</a></h3>
<p>For regular expression matching method, sangeranalyseR will group reads based on their contig name and read direction in their read names automatically; therefore, users have to follow the read-naming regulations below:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>All reads in the same contig group must include the same contig name in their read names.</p></li>
<li><p>Forward or reverse direction also has to be specified in their read names.</p></li>
</ul>
</div>
<p>There are three parameters, <code class="code docutils literal notranslate"><span class="pre">FASTA_File</span></code>, <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixForward</span></code> and <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixReverse</span></code>, that define the grouping rule to let sangeranalyseR automatically match correct reads in <strong>FASTA</strong> file and divide them into forward and reverse directions.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">FASTA_File</span></code>: this is the path to <strong>FASTA</strong> file that contains all sequences of reads, and it can be either an absolute or relative path. We suggest users to include only target reads inside this <strong>FASTA</strong> file and do not include any other unrelated reads.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixForward</span></code>: this is a regular expression that matches all read names in forward direction. <code class="code docutils literal notranslate"><span class="pre">grepl</span></code> function in R is used.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixReverse</span></code>: this is a regular expression that matches all read names in reverse direction. <code class="code docutils literal notranslate"><span class="pre">grepl</span></code> function in R is used.</p></li>
</ul>
</div>
<p>If you don't know what regular expression is, don't panic - it's just a way of recognising text. Please refer to <a class="reference internal" href="how_to_page.html#what-is-a-regular-expression"><span class="std std-ref">What is a regular expression?</span></a> for more details. Here is an example of how it works in sangeranalseR:</p>
<p>So how sangeranalyseR works is that it first matches the forward and reverse reads by matching <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixForward</span></code> and <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixReverse</span></code>. Then, sangeranalyseR uses the <code class="code docutils literal notranslate"><span class="pre">str_split</span></code> function to split and vectorize their read names into &quot;contig name&quot; and &quot;direction-suffix&quot; two parts. For those having the same &quot;contig name&quot; will be grouped into the same contig.</p>
<p>Therefore, it is important to have a consistent naming strategy. You need to make sure that reads in the <strong>FASTA</strong> file that are in the same contig group share the same contig name and carefully select your <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixForward</span></code> and <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixReverse</span></code>. The bad file-naming and wrong regex matching might accidentally include reverse reads into the forward read list or vice versa, which will make the program generate wrong results. So, how should we systematically name the reads? We suggest users to follow the file-naming regulation in <a class="reference internal" href="#sangeranalyser-filename-convention-sangeralignment-fasta"><span class="std std-ref">Figure_3</span></a>.</p>
<div class="figure align-center" id="id7">
<span id="sangeranalyser-filename-convention-sangeralignment-fasta"></span><a class="reference internal image-reference" href="../_images/sangeranalyseR_filename_convention_fasta.png"><img alt="../_images/sangeranalyseR_filename_convention_fasta.png" src="../_images/sangeranalyseR_filename_convention_fasta.png" style="width: 538.25px; height: 28.0px;" /></a>
<p class="caption"><span class="caption-text">Figure 3. Suggested read naming regulation in <strong>FASTA</strong> file - <em>SangerAlignment</em>.</span><a class="headerlink" href="#id7" title="Permalink to this image">¶</a></p>
</div>
<p>As you can see, the first part of the regulation is a consensus read name (or contig name), which helps sangeranalseR to identify which reads should be grouped into the same contig automatically. The second part of the regulation is an index; since there might be more than one read that is in the forward or reverse direction, we recommend you to number your reads in the same contig group. The Last part is a direction which is either 'F' (forward) or 'R' (reverse).</p>
<p>To make it more specific, let's go back to the true example. In <a class="reference internal" href="#sangeralignment-fasta-input"><span class="std std-ref">Figure_2</span></a>, there are eight reads in the <code class="code docutils literal notranslate"><span class="pre">FASTA</span></code> file (<code class="code docutils literal notranslate"><span class="pre">fasta_FN</span></code>; <a class="reference download internal" download="" href="../_downloads/bfd858bb8d73620d1300d2e3b2a34899/Sanger_all_reads.fa"><code class="xref download docutils literal notranslate"><span class="pre">Sanger_all_reads.fa</span></code></a>). First, we set <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixForward</span></code> to <code class="code docutils literal notranslate"><span class="pre">&quot;_[0-9]*_F$&quot;</span></code> and <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixReverse</span></code> to <code class="code docutils literal notranslate"><span class="pre">&quot;_[0-9]*_R$&quot;</span></code> to let sangeranalyseR match and group forward and reverse reads automatically. By the regular expression rule, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO006-09_1_F</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO007-09_1_F</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO040-09_1_F</span></code>, and <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO041-09_1_F</span></code>, are categorized into forward reads, and <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO006-09_1_R</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO007-09_1_R</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO040-09_1_R</span></code>, and <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO041-09_1_R</span></code> are categorized into reverse reads. Then, <code class="code docutils literal notranslate"><span class="pre">str_split</span></code> function is used to split each filename above into &quot;contig name&quot; and &quot;direction-suffix&quot;. Four contig names are detected in this example which are <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO006-09</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO007-09</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO040-09</span></code>, and <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO041-09</span></code>. Last, a loop iterates through all contig names, and sangeranalseR creates each of them into a <strong>SangerContig</strong> instance. You can check <a class="reference internal" href="advance_user_SangerContig_fasta.html#advanced-user-guide-sangercontig-fasta"><span class="std std-ref">Advanced User Guide - SangerContig (FASTA)</span></a> to see how sangeranalyseR creates a <em>SangerContig</em> instance.</p>
<p>The reason why we strongly recommend you to follow this file-naming regulation is that by doing so, you can directly adopt the example regular expression matching values, <code class="code docutils literal notranslate"><span class="pre">&quot;_[0-9]*_F$&quot;</span></code> and <code class="code docutils literal notranslate"><span class="pre">&quot;_[0-9]*_R$&quot;</span></code>, to group reads and reduce chances of error. Everything mentioned above will be done automatically.</p>
<p>After understanding how parameters work, please refer to <a class="reference internal" href="#creating-sangeralignment-instance-from-fasta"><span class="std std-ref">Creating SangerAlignment instance from FASTA</span></a> below to see how sangeranalseR creates <em>SangerAlignment</em> instance.</p>
</div>
<div class="section" id="csv-file-matching-sangeralignment-inputs-fasta">
<h3>(2) &quot;CSV file matching&quot; <em>SangerAlignment</em> inputs (<strong>FASTA</strong>)<a class="headerlink" href="#csv-file-matching-sangeralignment-inputs-fasta" title="Permalink to this headline">¶</a></h3>
<p>No doubt that read names in the original <strong>FASTA</strong> file do not follow the naming regulation, and you do not want to change the original <strong>FASTA</strong> file; thus, we provide a second grouping approach, CSV file matching method. sangeranalyseR will group reads in the <strong>FASTA</strong> file based on the information in a CSV file automatically, and users do not need to alter the read names in the <strong>FASTA</strong> file. The note below shows the regulations:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Here is an <a class="reference download internal" download="" href="../_downloads/76dbb0692fb55b06c60aaf559983d503/names_conversion.csv"><code class="xref download docutils literal notranslate"><span class="pre">example</span> <span class="pre">CSV</span> <span class="pre">file</span></code></a> (<a class="reference internal" href="#sangeranalyser-csv-file-sangeralignment-fasta"><span class="std std-ref">Figure 4</span></a>)</p>
<blockquote>
<div><div class="figure align-center" id="id8">
<span id="sangeranalyser-csv-file-sangeralignment-fasta"></span><a class="reference internal image-reference" href="../_images/sangeranalyseR_csv_file_sangeralignment_fasta.png"><img alt="../_images/sangeranalyseR_csv_file_sangeralignment_fasta.png" src="../_images/sangeranalyseR_csv_file_sangeralignment_fasta.png" style="width: 497.6px; height: 190.4px;" /></a>
<p class="caption"><span class="caption-text">Figure 4. Example CSV file for <em>SangerAlignment</em> instance creation.</span><a class="headerlink" href="#id8" title="Permalink to this image">¶</a></p>
</div>
</div></blockquote>
<ul class="simple">
<li><p>There must be three columns, &quot;<strong>reads</strong>&quot;, &quot;<strong>direction</strong>&quot;, and &quot;<strong>contig</strong>&quot;, in the CSV file.</p></li>
<li><p>The &quot;<strong>reads</strong>&quot; column stores the filename of <strong>AB1</strong> files that are going to be included in the analysis.</p></li>
<li><p>The &quot;<strong>direction</strong>&quot; column stores the direction of the reads. It must be &quot;F&quot; (forward) or &quot;R&quot; (reverse).</p></li>
<li><p>The &quot;<strong>contig</strong>&quot; column stores the contig name that each read blongs. Reads in the same contig have to have the same contig name, and they will be grouped into the same contig.</p></li>
</ul>
</div>
<p>There are two parameters, <code class="code docutils literal notranslate"><span class="pre">FASTA_File</span></code> and <code class="code docutils literal notranslate"><span class="pre">CSV_NamesConversion</span></code>,that define the grouping rule to help sangeranalseR to automatically match correct reads in the <strong>FASTA</strong> file and divide them into forward and reverse directions.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">FASTA_File</span></code>: this is the path to <strong>FASTA</strong> file that contains all sequences of reads, and it can be either an absolute or relative path. We suggest users to include only target reads inside this <strong>FASTA</strong> file and do not include any other unrelated reads.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">CSV_NamesConversion</span></code>: this is the path to the CSV file. It can be either an absolute or relative path.</p></li>
</ul>
</div>
<p>The main difference between &quot;CSV file matching&quot; and &quot;regular expression matching&quot; is where the grouping rule is written. For &quot;regular expression matching&quot;, rules are writtein in read names, and thus more naming requirements are required. In contrast, rules of &quot;CSV file matching&quot; are written in an additional CSV file so it is more flexible on naming reads.</p>
<p>So how sangeranalyseR works is that it first reads in the CSV file (with <em>&quot;reads&quot;</em>, <em>&quot;direction&quot;</em>, and <em>&quot;contig&quot;</em> columns), find the read names in the <strong>FASTA</strong> file that are listed in <em>&quot;reads&quot;</em>, and assign directions to them based on <em>&quot;direction&quot;</em>.</p>
<p>To make it more specific, let's go back to the true example. First, we prepare a <a class="reference download internal" download="" href="../_downloads/76dbb0692fb55b06c60aaf559983d503/names_conversion.csv"><code class="xref download docutils literal notranslate"><span class="pre">CSV</span> <span class="pre">file</span></code></a> (<code class="code docutils literal notranslate"><span class="pre">CSV_NamesConversion</span></code>) and a <a class="reference download internal" download="" href="../_downloads/bfd858bb8d73620d1300d2e3b2a34899/Sanger_all_reads.fa"><code class="xref download docutils literal notranslate"><span class="pre">fasta</span> <span class="pre">file</span></code></a> (<code class="code docutils literal notranslate"><span class="pre">FASTA_File</span></code>). In the CSV file, there are 8 rows and 4 distinct contig names. sangeranalyseR matches <em>&quot;reads&quot;</em> of these 8 rows to read names in the <strong>FASTA</strong> file. Then sangeranalyseR groups all matched reads, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO006-09_1_F</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO007-09_1_F</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO040-09_1_F</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO041-09_1_F</span></code>,   <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO006-09_1_R</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO007-09_1_R</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO040-09_1_R</span></code>, and <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO041-09_1_R</span></code>, into 4 distinct contigs which are <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO006-09</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO007-09</span></code>, <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO040-09</span></code>, and <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO041-09</span></code>, by the <em>&quot;contig&quot;</em> column. Last, the directions of reads in each contig are assigned by the <em>&quot;direction&quot;</em> column. Take <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO041-09</span></code> contig as an example. Its &quot;forward read list&quot; will include <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO041-09_1_F</span></code>, and its &quot;reverse read list&quot; will include <code class="code docutils literal notranslate"><span class="pre">Achl_ACHLO041-09_1_R</span></code>.</p>
<p>After understanding how parameters work, please refer to <a class="reference internal" href="#creating-sangeralignment-instance-from-fasta"><span class="std std-ref">Creating SangerAlignment instance from FASTA</span></a> below to see how sangeranalseR creates <em>SangerAlignment</em> instance.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
</div>
<div class="section" id="creating-sangeralignment-instance-from-fasta">
<h2>Creating <em>SangerAlignment</em> instance from <strong>FASTA</strong><a class="headerlink" href="#creating-sangeralignment-instance-from-fasta" title="Permalink to this headline">¶</a></h2>
<p>After preparing the input directory, we can create a <em>SangerAlignment</em> instance by running <code class="code docutils literal notranslate"><span class="pre">SangerAlignment</span></code> constructor function or <code class="code docutils literal notranslate"><span class="pre">new</span></code> method. The constructor function is a wrapper for <code class="code docutils literal notranslate"><span class="pre">new</span></code> method and it makes instance creation more intuitive. Their input parameters are same, and all of them have their default values. For more details about <em>SangerAlignment</em> inputs and slots definition, please refer to <a class="reference external" href="https://bioconductor.org/packages/release/bioc/manuals/sangeranalyseR/man/sangeranalyseR.pdf">sangeranalyseR reference manual</a>. We will explain two <em>SangerAlignment</em> instance creation methods, &quot;regular expression matching&quot; and &quot;CSV file matching&quot;.</p>
<div class="section" id="regular-expression-matching-sangeralignment-creation-fasta">
<h3>(1) &quot;regular expression matching&quot; <em>SangerAlignment</em> creation (<strong>FASTA</strong>)<a class="headerlink" href="#regular-expression-matching-sangeralignment-creation-fasta" title="Permalink to this headline">¶</a></h3>
<p>The consturctor function and <code class="code docutils literal notranslate"><span class="pre">new</span></code> method below contain three parameters, <code class="code docutils literal notranslate"><span class="pre">FASTA_File</span></code>, <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixForward</span></code>, and <code class="code docutils literal notranslate"><span class="pre">REGEX_SuffixReverse</span></code>, that we mentioned in the previous section. In contrast to <strong>AB1</strong> input method, it does not include quality trimming and chromatogram visualization parameters. Run the following code and create <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa</span></code> instance.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># using `constructor` function to create SangerAlignment instance</span>
<span class="n">my_sangerAlignmentFa</span> <span class="o">&lt;-</span> <span class="nf">SangerAlignment</span><span class="p">(</span><span class="n">inputSource</span>          <span class="o">=</span> <span class="s">&quot;FASTA&quot;</span><span class="p">,</span>
                                        <span class="n">processMethod</span>        <span class="o">=</span> <span class="s">&quot;REGEX&quot;</span><span class="p">,</span>
                                        <span class="n">FASTA_File</span>           <span class="o">=</span> <span class="n">fastaFN</span><span class="p">,</span>
                                        <span class="n">REGEX_SuffixForward</span>  <span class="o">=</span> <span class="s">&quot;_[0-9]*_F$&quot;</span><span class="p">,</span>
                                        <span class="n">REGEX_SuffixReverse</span>  <span class="o">=</span> <span class="s">&quot;_[0-9]*_R$&quot;</span><span class="p">,</span>
                                        <span class="n">refAminoAcidSeq</span>      <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">,</span>
                                        <span class="n">minReadsNum</span>           <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
                                        <span class="n">minReadLength</span>         <span class="o">=</span> <span class="m">20</span><span class="p">,</span>
                                        <span class="n">minFractionCall</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                                        <span class="n">maxFractionLost</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                                        <span class="n">geneticCode</span>           <span class="o">=</span> <span class="n">GENETIC_CODE</span><span class="p">,</span>
                                        <span class="n">acceptStopCodons</span>      <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                                        <span class="n">readingFrame</span>          <span class="o">=</span> <span class="m">1</span><span class="p">,</span>
                                        <span class="n">processorsNum</span>         <span class="o">=</span> <span class="m">1</span><span class="p">)</span>


<span class="n">my_sangerAlignmentFa</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;SangerAlignment&quot;</span><span class="p">,</span>
                            <span class="n">inputSource</span>          <span class="o">=</span> <span class="s">&quot;FASTA&quot;</span><span class="p">,</span>
                            <span class="n">processMethod</span>        <span class="o">=</span> <span class="s">&quot;REGEX&quot;</span><span class="p">,</span>
                            <span class="n">FASTA_File</span>           <span class="o">=</span> <span class="n">fastaFN</span><span class="p">,</span>
                            <span class="n">REGEX_SuffixForward</span>  <span class="o">=</span> <span class="s">&quot;_[0-9]*_F$&quot;</span><span class="p">,</span>
                            <span class="n">REGEX_SuffixReverse</span>  <span class="o">=</span> <span class="s">&quot;_[0-9]*_R$&quot;</span><span class="p">,</span>
                            <span class="n">refAminoAcidSeq</span>      <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">,</span>
                            <span class="n">minReadsNum</span>           <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
                            <span class="n">minReadLength</span>         <span class="o">=</span> <span class="m">20</span><span class="p">,</span>
                            <span class="n">minFractionCall</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                            <span class="n">maxFractionLost</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                            <span class="n">geneticCode</span>           <span class="o">=</span> <span class="n">GENETIC_CODE</span><span class="p">,</span>
                            <span class="n">acceptStopCodons</span>      <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                            <span class="n">readingFrame</span>          <span class="o">=</span> <span class="m">1</span><span class="p">,</span>
                            <span class="n">processorsNum</span>         <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
<p>In this example, 8 reads are detected and 4 distinct <em>SangerContig</em> instances are created. These <em>SangerContig</em> instances are stored in a &quot;contig list&quot; in <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa</span></code>, which will be used as the input for the following functions.</p>
<p>Inside the R shell, you can run <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa</span></code> to get basic information of the instance or run <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa&#64;objectResults&#64;readResultTable</span></code> to check the creation result of every Sanger read after <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa</span></code> is successfully created.</p>
<p>Here is the output of <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>SangerAlignment S4 instance
         Input Source :  FASTA
         Process Method :  REGEX
      Fasta File Name :  /Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata/fasta/SangerAlignment/Sanger_all_reads.fa
   REGEX Suffix Forward :  _[0-9]*_F$
   REGEX Suffix Reverse :  _[0-9]*_R$
      Contigs Consensus :  TTATAYTTTATTYTRGGCGTCTGAGCAGGAATGGTTGGAGCYGGTATAAGACTYCTAATTCGAATYGAGCTAAGACARCCRGGAGCRTTCCTRGGMAGRGAYCAACTMTAYAATACTATYGTWACTGCWCACGCATTTGTAATAATYTTCTTTCTAGTAATRCCTGTATTYATYGGGGGRTTCGGWAAYTGRCTTYTACCTTTAATACTTGGAGCCCCYGAYATRGCATTCCCWCGACTYAACAACATRAGATTCTGACTMCTTCCCCCATCACTRATCCTTYTAGTGTCCTCTGCKGCRGTAGAAAAAGGCGCTGGWACKGGRTGAACTGTTTATCCGCCYCTAGCAAGAAATMTTGCYCAYGCMGGCCCRTCTGTAGAYTTAGCYATYTTTTCYCTTCATTTAGCGGGTGCKTCWTCWATYYTAGGGGCYATTAATTTTATYACYACWGTTATTAAYATGCGWTGAAGAGGMTTACGWCTTGAACGAATYCCMYTRTTYGTYTGAGCYGTRCTAATTACAGTKGTTCTTCTACTYCTATCYTTACCAGTGYTAGCMGGTGCMATTACYATACTWCTTACCGAYCGAAAYCTCAATACYTCMTTCTTTGATCCTGCYGGTGGTGGAGAYCCCATCCTCTACCAACACTTATTCTGATTTTTTGGTCACCCTGAG
SUCCESS [2021-14-07 04:33:57] &#39;SangerAlignment&#39; is successfully created!
</pre></div>
</div>
<p>Here is the output of <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa&#64;objectResults&#64;readResultTable</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>            <span class="n">readName</span> <span class="n">creationResult</span> <span class="n">errorType</span> <span class="n">errorMessage</span> <span class="n">inputSource</span>    <span class="n">direction</span>
<span class="mi">1</span> <span class="n">Achl_ACHLO006</span><span class="o">-</span><span class="mi">09_1</span><span class="n">_F</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Forward</span> <span class="n">Read</span>
<span class="mi">2</span> <span class="n">Achl_ACHLO006</span><span class="o">-</span><span class="mi">09_2</span><span class="n">_R</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Reverse</span> <span class="n">Read</span>
<span class="mi">3</span> <span class="n">Achl_ACHLO007</span><span class="o">-</span><span class="mi">09_1</span><span class="n">_F</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Forward</span> <span class="n">Read</span>
<span class="mi">4</span> <span class="n">Achl_ACHLO007</span><span class="o">-</span><span class="mi">09_2</span><span class="n">_R</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Reverse</span> <span class="n">Read</span>
<span class="mi">5</span> <span class="n">Achl_ACHLO040</span><span class="o">-</span><span class="mi">09_1</span><span class="n">_F</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Forward</span> <span class="n">Read</span>
<span class="mi">6</span> <span class="n">Achl_ACHLO040</span><span class="o">-</span><span class="mi">09_2</span><span class="n">_R</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Reverse</span> <span class="n">Read</span>
<span class="mi">7</span> <span class="n">Achl_ACHLO041</span><span class="o">-</span><span class="mi">09_1</span><span class="n">_F</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Forward</span> <span class="n">Read</span>
<span class="mi">8</span> <span class="n">Achl_ACHLO041</span><span class="o">-</span><span class="mi">09_2</span><span class="n">_R</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Reverse</span> <span class="n">Read</span>
</pre></div>
</div>
</div>
<div class="section" id="csv-file-matching-sangeralignment-creation-fasta">
<h3>(2) &quot;CSV file matching&quot; <em>SangerAlignment</em> creation (<strong>FASTA</strong>)<a class="headerlink" href="#csv-file-matching-sangeralignment-creation-fasta" title="Permalink to this headline">¶</a></h3>
<p>The consturctor function and <code class="code docutils literal notranslate"><span class="pre">new</span></code> method below contain two parameters, <code class="code docutils literal notranslate"><span class="pre">FASTA_File</span></code>, and <code class="code docutils literal notranslate"><span class="pre">CSV_NamesConversion</span></code>, that we mentioned in the previous section. Run the following code and create <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa</span></code> instance.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">csv_namesConversion</span> <span class="o">&lt;-</span> <span class="nf">file.path</span><span class="p">(</span><span class="n">rawDataDir</span><span class="p">,</span> <span class="s">&quot;fasta&quot;</span><span class="p">,</span> <span class="s">&quot;SangerAlignment&quot;</span><span class="p">,</span> <span class="s">&quot;names_conversion.csv&quot;</span><span class="p">)</span>

<span class="c1"># using `constructor` function to create SangerAlignment instance</span>
<span class="n">my_sangerAlignmentFa</span> <span class="o">&lt;-</span> <span class="nf">SangerAlignment</span><span class="p">(</span><span class="n">inputSource</span>          <span class="o">=</span> <span class="s">&quot;FASTA&quot;</span><span class="p">,</span>
                                        <span class="n">processMethod</span>        <span class="o">=</span> <span class="s">&quot;CSV&quot;</span><span class="p">,</span>
                                        <span class="n">FASTA_File</span>           <span class="o">=</span> <span class="n">fastaFN</span><span class="p">,</span>
                                        <span class="n">CSV_NamesConversion</span>  <span class="o">=</span> <span class="n">csv_namesConversion</span><span class="p">,</span>
                                        <span class="n">refAminoAcidSeq</span>      <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">,</span>
                                        <span class="n">minReadsNum</span>           <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
                                        <span class="n">minReadLength</span>         <span class="o">=</span> <span class="m">20</span><span class="p">,</span>
                                        <span class="n">minFractionCall</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                                        <span class="n">maxFractionLost</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                                        <span class="n">geneticCode</span>           <span class="o">=</span> <span class="n">GENETIC_CODE</span><span class="p">,</span>
                                        <span class="n">acceptStopCodons</span>      <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                                        <span class="n">readingFrame</span>          <span class="o">=</span> <span class="m">1</span><span class="p">,</span>
                                        <span class="n">processorsNum</span>         <span class="o">=</span> <span class="m">1</span><span class="p">)</span>


<span class="n">my_sangerAlignmentFa</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;SangerAlignment&quot;</span><span class="p">,</span>
                            <span class="n">inputSource</span>          <span class="o">=</span> <span class="s">&quot;FASTA&quot;</span><span class="p">,</span>
                            <span class="n">processMethod</span>        <span class="o">=</span> <span class="s">&quot;CSV&quot;</span><span class="p">,</span>
                            <span class="n">FASTA_File</span>           <span class="o">=</span> <span class="n">fastaFN</span><span class="p">,</span>
                            <span class="n">CSV_NamesConversion</span>  <span class="o">=</span> <span class="n">csv_namesConversion</span><span class="p">,</span>
                            <span class="n">refAminoAcidSeq</span>      <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">,</span>
                            <span class="n">minReadsNum</span>           <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
                            <span class="n">minReadLength</span>         <span class="o">=</span> <span class="m">20</span><span class="p">,</span>
                            <span class="n">minFractionCall</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                            <span class="n">maxFractionLost</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                            <span class="n">geneticCode</span>           <span class="o">=</span> <span class="n">GENETIC_CODE</span><span class="p">,</span>
                            <span class="n">acceptStopCodons</span>      <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                            <span class="n">readingFrame</span>          <span class="o">=</span> <span class="m">1</span><span class="p">,</span>
                            <span class="n">processorsNum</span>         <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
<p>First, you need to load the CSV file into the R environment. If you are still don't know how to prepare it, please check <a class="reference internal" href="#csv-file-matching-sangeralignment-inputs-fasta"><span class="std std-ref">(2) &quot;CSV file matching&quot; SangerAlignment inputs (FASTA)</span></a>. Then, it will follow rules in the CSV file and create <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa</span></code>. After it's created, inside the R shell, you can run <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa</span></code> to get basic information of the instance or run <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa&#64;objectResults&#64;readResultTable</span></code> to check the creation result of every Sanger read after <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa</span></code> is successfully created.</p>
<p>Here is the output of <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>SangerAlignment S4 instance
         Input Source :  FASTA
         Process Method :  CSV
      Fasta File Name :  /Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata/fasta/SangerAlignment/Sanger_all_reads.fa
   CSV Names Conversion :  /Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata/fasta/SangerAlignment/names_conversion.csv
      Contigs Consensus :  TTATAYTTTATTYTRGGCGTCTGAGCAGGAATGGTTGGAGCYGGTATAAGACTYCTAATTCGAATYGAGCTAAGACARCCRGGAGCRTTCCTRGGMAGRGAYCAACTMTAYAATACTATYGTWACTGCWCACGCATTTGTAATAATYTTCTTTCTAGTAATRCCTGTATTYATYGGGGGRTTCGGWAAYTGRCTTYTACCTTTAATACTTGGAGCCCCYGAYATRGCATTCCCWCGACTYAACAACATRAGATTCTGACTMCTTCCCCCATCACTRATCCTTYTAGTGTCCTCTGCKGCRGTAGAAAAAGGCGCTGGWACKGGRTGAACTGTTTATCCGCCYCTAGCAAGAAATMTTGCYCAYGCMGGCCCRTCTGTAGAYTTAGCYATYTTTTCYCTTCATTTAGCGGGTGCKTCWTCWATYYTAGGGGCYATTAATTTTATYACYACWGTTATTAAYATGCGWTGAAGAGGMTTACGWCTTGAACGAATYCCMYTRTTYGTYTGAGCYGTRCTAATTACAGTKGTTCTTCTACTYCTATCYTTACCAGTGYTAGCMGGTGCMATTACYATACTWCTTACCGAYCGAAAYCTCAATACYTCMTTCTTTGATCCTGCYGGTGGTGGAGAYCCCATCCTCTACCAACACTTATTCTGATTTTTTGGTCACCCTGAG
SUCCESS [2021-14-07 04:38:44] &#39;SangerAlignment&#39; is successfully created!
</pre></div>
</div>
<p>Here is the output of <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa&#64;objectResults&#64;readResultTable</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>            <span class="n">readName</span> <span class="n">creationResult</span> <span class="n">errorType</span> <span class="n">errorMessage</span> <span class="n">inputSource</span>    <span class="n">direction</span>
<span class="mi">1</span> <span class="n">Achl_ACHLO006</span><span class="o">-</span><span class="mi">09_1</span><span class="n">_F</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Forward</span> <span class="n">Read</span>
<span class="mi">2</span> <span class="n">Achl_ACHLO006</span><span class="o">-</span><span class="mi">09_2</span><span class="n">_R</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Reverse</span> <span class="n">Read</span>
<span class="mi">3</span> <span class="n">Achl_ACHLO007</span><span class="o">-</span><span class="mi">09_1</span><span class="n">_F</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Forward</span> <span class="n">Read</span>
<span class="mi">4</span> <span class="n">Achl_ACHLO007</span><span class="o">-</span><span class="mi">09_2</span><span class="n">_R</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Reverse</span> <span class="n">Read</span>
<span class="mi">5</span> <span class="n">Achl_ACHLO040</span><span class="o">-</span><span class="mi">09_1</span><span class="n">_F</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Forward</span> <span class="n">Read</span>
<span class="mi">6</span> <span class="n">Achl_ACHLO040</span><span class="o">-</span><span class="mi">09_2</span><span class="n">_R</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Reverse</span> <span class="n">Read</span>
<span class="mi">7</span> <span class="n">Achl_ACHLO041</span><span class="o">-</span><span class="mi">09_1</span><span class="n">_F</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Forward</span> <span class="n">Read</span>
<span class="mi">8</span> <span class="n">Achl_ACHLO041</span><span class="o">-</span><span class="mi">09_2</span><span class="n">_R</span>           <span class="n">TRUE</span>      <span class="kc">None</span>         <span class="kc">None</span>       <span class="n">FASTA</span> <span class="n">Reverse</span> <span class="n">Read</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
</div>
<div class="section" id="writing-sangeralignment-fasta-files-fasta">
<h2>Writing <em>SangerAlignment</em> FASTA files <sub>(FASTA)</sub><a class="headerlink" href="#writing-sangeralignment-fasta-files-fasta" title="Permalink to this headline">¶</a></h2>
<p>Users can write the <em>SangerAlignment</em> instance, <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa</span></code>, to <strong>FASTA</strong> files. There are four options for users to choose from in <code class="code docutils literal notranslate"><span class="pre">selection</span></code> parameter.</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">reads_unalignment</span></code>: Writing reads into a single <strong>FASTA</strong> file (only trimmed without alignment).</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">reads_alignment</span></code>: Writing reads alignment and contig read to a single <strong>FASTA</strong> file.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">contig</span></code>: Writing the contig to a single <strong>FASTA</strong> file.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">all</span></code>: Writing reads, reads alignment, and the contig into three different files.</p></li>
</ul>
<p>Below is the oneliner for writing out <strong>FASTA</strong> files. This function mainly depends on <code class="code docutils literal notranslate"><span class="pre">writeXStringSet</span></code> function in <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/Biostrings.html">Biostrings</a> R package. Users can set the compression level through <code class="code docutils literal notranslate"><span class="pre">writeFasta</span></code> function.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">writeFasta</span><span class="p">(</span><span class="n">my_sangerAlignmentFa</span><span class="p">,</span>
           <span class="n">outputDir</span>         <span class="o">=</span> <span class="nf">tempdir</span><span class="p">(),</span>
           <span class="n">compress</span>          <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>
           <span class="n">compression_level</span> <span class="o">=</span> <span class="kc">NA</span><span class="p">,</span>
           <span class="n">selection</span>         <span class="o">=</span> <span class="s">&quot;all&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Users can download the output FASTA file of this example through the following three links:</p>
<ol class="arabic simple">
<li><p><a class="reference download internal" download="" href="../_downloads/165af37e8893cfdfa4dba91c44cd791b/Sanger_contigs_unalignment.fa"><code class="xref download docutils literal notranslate"><span class="pre">Sanger_contigs_unalignment.fa</span></code></a></p></li>
<li><p><a class="reference download internal" download="" href="../_downloads/144d09562ae5b2c7dc079e4bc2b33fb7/Sanger_contigs_alignment.fa"><code class="xref download docutils literal notranslate"><span class="pre">Sanger_contigs_alignment.fa</span></code></a></p></li>
<li><p><a class="reference download internal" download="" href="../_downloads/81a7338101f110748cb72056d9975225/Sanger_all_trimmed_reads.fa"><code class="xref download docutils literal notranslate"><span class="pre">Sanger_all_trimmed_reads.fa</span></code></a></p></li>
</ol>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="generating-sangeralignment-report-fasta">
<h2>Generating <em>SangerAlignment</em> report <sub>(FASTA)</sub><a class="headerlink" href="#generating-sangeralignment-report-fasta" title="Permalink to this headline">¶</a></h2>
<p>Last but not least, users can save <em>SangerAlignment</em> instance, <code class="code docutils literal notranslate"><span class="pre">my_sangerAlignmentFa</span></code>, into a report after the analysis. The report will be generated in <strong>HTML</strong> by knitting <strong>Rmd</strong> files.</p>
<p>Users can set <code class="code docutils literal notranslate"><span class="pre">includeSangerContig</span></code> and <code class="code docutils literal notranslate"><span class="pre">includeSangerRead</span></code> parameters to decide to which level the <em>SangerAlignment</em> report will go. Moreover, after the reports are generated, users can easily navigate through reports in different levels within the <strong>HTML</strong> file.</p>
<p>One thing to pay attention to is that if users have many reads, it will take quite a long time to write out all reports. If users only want to generate the contig result, remember to set <code class="code docutils literal notranslate"><span class="pre">includeSangerRead</span></code> and <code class="code docutils literal notranslate"><span class="pre">includeSangerContig</span></code> to <code class="code docutils literal notranslate"><span class="pre">FALSE</span></code> in order to save time.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">generateReport</span><span class="p">(</span><span class="n">my_sangerAlignmentFa</span><span class="p">,</span>
               <span class="n">outputDir</span>           <span class="o">=</span> <span class="nf">tempdir</span><span class="p">(),</span>
               <span class="n">includeSangerRead</span>   <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>
               <span class="n">includeSangerContig</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
</pre></div>
</div>
<p>Here is the generated <a class="reference external" href="https://kuanhao-chao.github.io/sangeranalyseR_report/SangerAlignment/FASTA/SangerAlignment/SangerAlignment_Report.html">SangerAlignment html report of this example (FASTA)</a>. Users can access to '<em>Basic Information</em>', '<em>Contigs Consensus</em>', '<em>Contigs Alignment</em>', '<em>Contigs Tree</em>', and '<em>Contig Reports</em>' sections inside it. Furthermore, users can also navigate through html reports of all contigs and forward and reverse <em>SangerRead</em> in this <em>SangerAlignment</em> report.</p>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="code-summary-sangeralignment-fasta">
<h2>Code summary (<em>SangerAlignment</em>, <strong>FASTA</strong>)<a class="headerlink" href="#code-summary-sangeralignment-fasta" title="Permalink to this headline">¶</a></h2>
<div class="section" id="preparing-sangeralignment-fasta-inputs">
<h3>(1) Preparing <em>SangerAlignment</em> <strong>FASTA</strong> inputs<a class="headerlink" href="#preparing-sangeralignment-fasta-inputs" title="Permalink to this headline">¶</a></h3>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">rawDataDir</span> <span class="o">&lt;-</span> <span class="nf">system.file</span><span class="p">(</span><span class="s">&quot;extdata&quot;</span><span class="p">,</span> <span class="n">package</span> <span class="o">=</span> <span class="s">&quot;sangeranalyseR&quot;</span><span class="p">)</span>
<span class="n">fastaFN</span> <span class="o">&lt;-</span> <span class="nf">file.path</span><span class="p">(</span><span class="n">rawDataDir</span><span class="p">,</span> <span class="s">&quot;fasta&quot;</span><span class="p">,</span> <span class="s">&quot;SangerAlignment&quot;</span><span class="p">,</span> <span class="s">&quot;Sanger_all_reads.fa&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="id1">
<h3>(2) Creating <em>SangerAlignment</em> instance from <strong>FASTA</strong><a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<div class="section" id="regular-expression-method-sangeralignment-creation-fasta">
<h4>(2.1) &quot;Regular Expression Method&quot; <em>SangerAlignment</em> creation (<strong>FASTA</strong>)<a class="headerlink" href="#regular-expression-method-sangeralignment-creation-fasta" title="Permalink to this headline">¶</a></h4>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># using `constructor` function to create SangerAlignment instance</span>
<span class="n">my_sangerAlignmentFa</span> <span class="o">&lt;-</span> <span class="nf">SangerAlignment</span><span class="p">(</span><span class="n">inputSource</span>          <span class="o">=</span> <span class="s">&quot;FASTA&quot;</span><span class="p">,</span>
                                        <span class="n">processMethod</span>        <span class="o">=</span> <span class="s">&quot;REGEX&quot;</span><span class="p">,</span>
                                        <span class="n">FASTA_File</span>           <span class="o">=</span> <span class="n">fastaFN</span><span class="p">,</span>
                                        <span class="n">REGEX_SuffixForward</span>  <span class="o">=</span> <span class="s">&quot;_[0-9]*_F$&quot;</span><span class="p">,</span>
                                        <span class="n">REGEX_SuffixReverse</span>  <span class="o">=</span> <span class="s">&quot;_[0-9]*_R$&quot;</span><span class="p">,</span>
                                        <span class="n">refAminoAcidSeq</span>      <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">)</span>


<span class="n">my_sangerAlignmentFa</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;SangerAlignment&quot;</span><span class="p">,</span>
                            <span class="n">inputSource</span>          <span class="o">=</span> <span class="s">&quot;FASTA&quot;</span><span class="p">,</span>
                            <span class="n">processMethod</span>        <span class="o">=</span> <span class="s">&quot;REGEX&quot;</span><span class="p">,</span>
                            <span class="n">FASTA_File</span>           <span class="o">=</span> <span class="n">fastaFN</span><span class="p">,</span>
                            <span class="n">REGEX_SuffixForward</span>  <span class="o">=</span> <span class="s">&quot;_[0-9]*_F$&quot;</span><span class="p">,</span>
                            <span class="n">REGEX_SuffixReverse</span>  <span class="o">=</span> <span class="s">&quot;_[0-9]*_R$&quot;</span><span class="p">,</span>
                            <span class="n">refAminoAcidSeq</span>      <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="toggle docutils container">
<div class="header docutils container">
<p>Following is the R shell output that you will get.</p>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>INFO [2021-14-07 04:51:42] #################################################
INFO [2021-14-07 04:51:42] #### Start creating SangerAlignment instance ####
INFO [2021-14-07 04:51:42] #################################################
INFO [2021-14-07 04:51:42] **** You are using Regular Expression Method to group reads in FASTA file (No CSV file)!
INFO [2021-14-07 04:51:42] ========================================================
INFO [2021-14-07 04:51:42] ================ Creating &#39;SangerContig&#39; ===============
INFO [2021-14-07 04:51:42] ========================================================
INFO [2021-14-07 04:51:42]   &gt;&gt; Contig Name: &#39;Achl_ACHLO006-09&#39;
SUCCESS [2021-14-07 04:51:43] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:43] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:51:43] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:43]    &gt;&gt; &#39;Achl_ACHLO006-09_1_F&#39; is created (Forward Read; FASTA).
SUCCESS [2021-14-07 04:51:43] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:43] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:51:43] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:43]    &gt;&gt; &#39;Achl_ACHLO006-09_2_R&#39; is created (Reverse Read; FASTA).
INFO [2021-14-07 04:51:43]    &gt;&gt; The number of reads detected: 2
Assessing frameshifts in nucleotide sequences:
|=================================================================================================| 100%

Time difference of 0.03 secs
SUCCESS [2021-14-07 04:51:45] ==========================================================
SUCCESS [2021-14-07 04:51:45] ======== &#39;SangerContig&#39; S4 instance is created !! ========
SUCCESS [2021-14-07 04:51:45] ==========================================================
INFO [2021-14-07 04:51:45]    &gt;&gt; 2 read(s) created from FASTA file.
INFO [2021-14-07 04:51:45]      &gt;&gt; 1 reads assigned to &#39;forward reads&#39; according to &#39;regular expression&#39;.
INFO [2021-14-07 04:51:45]      &gt;&gt; 1 reads assigned to &#39;reverse reads&#39; according to &#39;regular expression&#39;.
INFO [2021-14-07 04:51:45] ========================================================
INFO [2021-14-07 04:51:45] ================ Creating &#39;SangerContig&#39; ===============
INFO [2021-14-07 04:51:45] ========================================================
INFO [2021-14-07 04:51:45]   &gt;&gt; Contig Name: &#39;Achl_ACHLO007-09&#39;
SUCCESS [2021-14-07 04:51:45] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:45] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:51:45] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:45]    &gt;&gt; &#39;Achl_ACHLO007-09_1_F&#39; is created (Forward Read; FASTA).
SUCCESS [2021-14-07 04:51:46] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:46] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:51:46] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:46]    &gt;&gt; &#39;Achl_ACHLO007-09_2_R&#39; is created (Reverse Read; FASTA).
INFO [2021-14-07 04:51:46]    &gt;&gt; The number of reads detected: 2
Assessing frameshifts in nucleotide sequences:
|=================================================================================================| 100%

Time difference of 0.03 secs
SUCCESS [2021-14-07 04:51:47] ==========================================================
SUCCESS [2021-14-07 04:51:48] ======== &#39;SangerContig&#39; S4 instance is created !! ========
SUCCESS [2021-14-07 04:51:48] ==========================================================
INFO [2021-14-07 04:51:48]    &gt;&gt; 2 read(s) created from FASTA file.
INFO [2021-14-07 04:51:48]      &gt;&gt; 1 reads assigned to &#39;forward reads&#39; according to &#39;regular expression&#39;.
INFO [2021-14-07 04:51:48]      &gt;&gt; 1 reads assigned to &#39;reverse reads&#39; according to &#39;regular expression&#39;.
INFO [2021-14-07 04:51:48] ========================================================
INFO [2021-14-07 04:51:48] ================ Creating &#39;SangerContig&#39; ===============
INFO [2021-14-07 04:51:48] ========================================================
INFO [2021-14-07 04:51:48]   &gt;&gt; Contig Name: &#39;Achl_ACHLO040-09&#39;
SUCCESS [2021-14-07 04:51:48] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:48] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:51:48] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:48]    &gt;&gt; &#39;Achl_ACHLO040-09_1_F&#39; is created (Forward Read; FASTA).
SUCCESS [2021-14-07 04:51:48] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:48] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:51:48] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:48]    &gt;&gt; &#39;Achl_ACHLO040-09_2_R&#39; is created (Reverse Read; FASTA).
INFO [2021-14-07 04:51:49]    &gt;&gt; The number of reads detected: 2
Assessing frameshifts in nucleotide sequences:
|=================================================================================================| 100%

Time difference of 0.03 secs
SUCCESS [2021-14-07 04:51:50] ==========================================================
SUCCESS [2021-14-07 04:51:50] ======== &#39;SangerContig&#39; S4 instance is created !! ========
SUCCESS [2021-14-07 04:51:50] ==========================================================
INFO [2021-14-07 04:51:50]    &gt;&gt; 2 read(s) created from FASTA file.
INFO [2021-14-07 04:51:50]      &gt;&gt; 1 reads assigned to &#39;forward reads&#39; according to &#39;regular expression&#39;.
INFO [2021-14-07 04:51:50]      &gt;&gt; 1 reads assigned to &#39;reverse reads&#39; according to &#39;regular expression&#39;.
INFO [2021-14-07 04:51:50] ========================================================
INFO [2021-14-07 04:51:50] ================ Creating &#39;SangerContig&#39; ===============
INFO [2021-14-07 04:51:50] ========================================================
INFO [2021-14-07 04:51:50]   &gt;&gt; Contig Name: &#39;Achl_ACHLO041-09&#39;
SUCCESS [2021-14-07 04:51:50] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:50] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:51:50] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:50]    &gt;&gt; &#39;Achl_ACHLO041-09_1_F&#39; is created (Forward Read; FASTA).
SUCCESS [2021-14-07 04:51:51] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:51] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:51:51] --------------------------------------------------------
SUCCESS [2021-14-07 04:51:51]    &gt;&gt; &#39;Achl_ACHLO041-09_2_R&#39; is created (Reverse Read; FASTA).
INFO [2021-14-07 04:51:51]    &gt;&gt; The number of reads detected: 2
Assessing frameshifts in nucleotide sequences:
|=================================================================================================| 100%

Time difference of 0.03 secs
SUCCESS [2021-14-07 04:51:52] ==========================================================
SUCCESS [2021-14-07 04:51:52] ======== &#39;SangerContig&#39; S4 instance is created !! ========
SUCCESS [2021-14-07 04:51:52] ==========================================================
INFO [2021-14-07 04:51:52]    &gt;&gt; 2 read(s) created from FASTA file.
INFO [2021-14-07 04:51:52]      &gt;&gt; 1 reads assigned to &#39;forward reads&#39; according to &#39;regular expression&#39;.
INFO [2021-14-07 04:51:52]      &gt;&gt; 1 reads assigned to &#39;reverse reads&#39; according to &#39;regular expression&#39;.
INFO [2021-14-07 04:51:52] Aligning consensus reads ...
SUCCESS [2021-14-07 04:51:53] #############################################################
SUCCESS [2021-14-07 04:51:53] ######## &#39;SangerAlignment&#39; S4 instance is created !! ########
SUCCESS [2021-14-07 04:51:53] #############################################################
INFO [2021-14-07 04:51:53]   &gt;&gt; 4 contigs detected from &#39;regular expression&#39;.
INFO [2021-14-07 04:51:53]       &gt;&gt; Contig &#39;Achl_ACHLO006-09&#39;:
INFO [2021-14-07 04:51:53]           &gt;&gt; 1 forward reads.
INFO [2021-14-07 04:51:53]           &gt;&gt; 1 reverse reads.
INFO [2021-14-07 04:51:53]       &gt;&gt; Contig &#39;Achl_ACHLO007-09&#39;:
INFO [2021-14-07 04:51:53]           &gt;&gt; 1 forward reads.
INFO [2021-14-07 04:51:53]           &gt;&gt; 1 reverse reads.
INFO [2021-14-07 04:51:53]       &gt;&gt; Contig &#39;Achl_ACHLO040-09&#39;:
INFO [2021-14-07 04:51:53]           &gt;&gt; 1 forward reads.
INFO [2021-14-07 04:51:53]           &gt;&gt; 1 reverse reads.
INFO [2021-14-07 04:51:53]       &gt;&gt; Contig &#39;Achl_ACHLO041-09&#39;:
INFO [2021-14-07 04:51:53]           &gt;&gt; 1 forward reads.
INFO [2021-14-07 04:51:53]           &gt;&gt; 1 reverse reads.
INFO [2021-14-07 04:51:53]   &gt;&gt; 8 reads created from FASTA file.
DEBUG [2021-14-07 04:51:53]    &gt;&gt; For more information, please run &#39;object&#39;.
DEBUG [2021-14-07 04:51:53]    &gt;&gt; Run &#39;object@objectResults@readResultTable&#39; to check the results of each Sanger reads
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="id2">
<h4>(2.2) &quot;CSV file matching&quot; <em>SangerAlignment</em> creation (<strong>FASTA</strong>)<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h4>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">csv_namesConversion</span> <span class="o">&lt;-</span> <span class="nf">file.path</span><span class="p">(</span><span class="n">rawDataDir</span><span class="p">,</span> <span class="s">&quot;fasta&quot;</span><span class="p">,</span> <span class="s">&quot;SangerAlignment&quot;</span><span class="p">,</span> <span class="s">&quot;names_conversion.csv&quot;</span><span class="p">)</span>

<span class="c1"># using `constructor` function to create SangerAlignment instance</span>
<span class="n">my_sangerAlignmentFa</span> <span class="o">&lt;-</span> <span class="nf">SangerAlignment</span><span class="p">(</span><span class="n">inputSource</span>          <span class="o">=</span> <span class="s">&quot;FASTA&quot;</span><span class="p">,</span>
                                        <span class="n">processMethod</span>        <span class="o">=</span> <span class="s">&quot;CSV&quot;</span><span class="p">,</span>
                                        <span class="n">FASTA_File</span>           <span class="o">=</span> <span class="n">fastaFN</span><span class="p">,</span>
                                        <span class="n">CSV_NamesConversion</span>  <span class="o">=</span> <span class="n">csv_namesConversion</span><span class="p">,</span>
                                        <span class="n">refAminoAcidSeq</span>      <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">)</span>


<span class="n">my_sangerAlignmentFa</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;SangerAlignment&quot;</span><span class="p">,</span>
                            <span class="n">inputSource</span>          <span class="o">=</span> <span class="s">&quot;FASTA&quot;</span><span class="p">,</span>
                            <span class="n">processMethod</span>        <span class="o">=</span> <span class="s">&quot;CSV&quot;</span><span class="p">,</span>
                            <span class="n">FASTA_File</span>           <span class="o">=</span> <span class="n">fastaFN</span><span class="p">,</span>
                            <span class="n">CSV_NamesConversion</span>  <span class="o">=</span> <span class="n">csv_namesConversion</span><span class="p">,</span>
`<span class="n">refAminoAcidSeq</span>      <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="toggle docutils container">
<div class="header docutils container">
<p>Following is the R shell output that you will get.</p>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>INFO [2021-14-07 04:52:53] #################################################
INFO [2021-14-07 04:52:53] #### Start creating SangerAlignment instance ####
INFO [2021-14-07 04:52:53] #################################################
INFO [2021-14-07 04:52:53] **** You are using CSV Name Conversion Method to group reads in FASTA file (with Csv file)!
INFO [2021-14-07 04:52:53] ========================================================
INFO [2021-14-07 04:52:53] ================ Creating &#39;SangerContig&#39; ===============
INFO [2021-14-07 04:52:53] ========================================================
INFO [2021-14-07 04:52:53]   &gt;&gt; Contig Name: &#39;Achl_ACHLO006-09&#39;
SUCCESS [2021-14-07 04:52:53] --------------------------------------------------------
SUCCESS [2021-14-07 04:52:53] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:52:53] --------------------------------------------------------
SUCCESS [2021-14-07 04:52:53]    &gt;&gt; &#39;Achl_ACHLO006-09_1_F&#39; is created (Forward Read; FASTA).
SUCCESS [2021-14-07 04:52:54] --------------------------------------------------------
SUCCESS [2021-14-07 04:52:54] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:52:54] --------------------------------------------------------
SUCCESS [2021-14-07 04:52:54]    &gt;&gt; &#39;Achl_ACHLO006-09_2_R&#39; is created (Reverse Read; FASTA).
INFO [2021-14-07 04:52:54]    &gt;&gt; The number of reads detected: 2
Assessing frameshifts in nucleotide sequences:
|=================================================================================================| 100%

Time difference of 0.04 secs
SUCCESS [2021-14-07 04:52:55] ==========================================================
SUCCESS [2021-14-07 04:52:55] ======== &#39;SangerContig&#39; S4 instance is created !! ========
SUCCESS [2021-14-07 04:52:55] ==========================================================
INFO [2021-14-07 04:52:55]    &gt;&gt; 2 read(s) created from FASTA file.
INFO [2021-14-07 04:52:55]      &gt;&gt; 1 reads assigned to &#39;forward reads&#39; according to &#39;csv file&#39;.
INFO [2021-14-07 04:52:55]      &gt;&gt; 1 reads assigned to &#39;reverse reads&#39; according to &#39;csv file&#39;.
INFO [2021-14-07 04:52:55] ========================================================
INFO [2021-14-07 04:52:55] ================ Creating &#39;SangerContig&#39; ===============
INFO [2021-14-07 04:52:55] ========================================================
INFO [2021-14-07 04:52:55]   &gt;&gt; Contig Name: &#39;Achl_ACHLO007-09&#39;
SUCCESS [2021-14-07 04:52:55] --------------------------------------------------------
SUCCESS [2021-14-07 04:52:55] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:52:55] --------------------------------------------------------
SUCCESS [2021-14-07 04:52:55]    &gt;&gt; &#39;Achl_ACHLO007-09_1_F&#39; is created (Forward Read; FASTA).
SUCCESS [2021-14-07 04:52:56] --------------------------------------------------------
SUCCESS [2021-14-07 04:52:56] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:52:56] --------------------------------------------------------
SUCCESS [2021-14-07 04:52:56]    &gt;&gt; &#39;Achl_ACHLO007-09_2_R&#39; is created (Reverse Read; FASTA).
INFO [2021-14-07 04:52:56]    &gt;&gt; The number of reads detected: 2
Assessing frameshifts in nucleotide sequences:
|=================================================================================================| 100%

Time difference of 0.05 secs
SUCCESS [2021-14-07 04:52:57] ==========================================================
SUCCESS [2021-14-07 04:52:57] ======== &#39;SangerContig&#39; S4 instance is created !! ========
SUCCESS [2021-14-07 04:52:57] ==========================================================
INFO [2021-14-07 04:52:58]    &gt;&gt; 2 read(s) created from FASTA file.
INFO [2021-14-07 04:52:58]      &gt;&gt; 1 reads assigned to &#39;forward reads&#39; according to &#39;csv file&#39;.
INFO [2021-14-07 04:52:58]      &gt;&gt; 1 reads assigned to &#39;reverse reads&#39; according to &#39;csv file&#39;.
INFO [2021-14-07 04:52:58] ========================================================
INFO [2021-14-07 04:52:58] ================ Creating &#39;SangerContig&#39; ===============
INFO [2021-14-07 04:52:58] ========================================================
INFO [2021-14-07 04:52:58]   &gt;&gt; Contig Name: &#39;Achl_ACHLO040-09&#39;
SUCCESS [2021-14-07 04:52:58] --------------------------------------------------------
SUCCESS [2021-14-07 04:52:58] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:52:58] --------------------------------------------------------
SUCCESS [2021-14-07 04:52:58]    &gt;&gt; &#39;Achl_ACHLO040-09_1_F&#39; is created (Forward Read; FASTA).
SUCCESS [2021-14-07 04:52:58] --------------------------------------------------------
SUCCESS [2021-14-07 04:52:58] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:52:58] --------------------------------------------------------
SUCCESS [2021-14-07 04:52:58]    &gt;&gt; &#39;Achl_ACHLO040-09_2_R&#39; is created (Reverse Read; FASTA).
INFO [2021-14-07 04:52:58]    &gt;&gt; The number of reads detected: 2
Assessing frameshifts in nucleotide sequences:
|=================================================================================================| 100%

Time difference of 0.03 secs
SUCCESS [2021-14-07 04:52:59] ==========================================================
SUCCESS [2021-14-07 04:53:00] ======== &#39;SangerContig&#39; S4 instance is created !! ========
SUCCESS [2021-14-07 04:53:00] ==========================================================
INFO [2021-14-07 04:53:00]    &gt;&gt; 2 read(s) created from FASTA file.
INFO [2021-14-07 04:53:00]      &gt;&gt; 1 reads assigned to &#39;forward reads&#39; according to &#39;csv file&#39;.
INFO [2021-14-07 04:53:00]      &gt;&gt; 1 reads assigned to &#39;reverse reads&#39; according to &#39;csv file&#39;.
INFO [2021-14-07 04:53:00] ========================================================
INFO [2021-14-07 04:53:00] ================ Creating &#39;SangerContig&#39; ===============
INFO [2021-14-07 04:53:00] ========================================================
INFO [2021-14-07 04:53:00]   &gt;&gt; Contig Name: &#39;Achl_ACHLO041-09&#39;
SUCCESS [2021-14-07 04:53:00] --------------------------------------------------------
SUCCESS [2021-14-07 04:53:00] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:53:00] --------------------------------------------------------
SUCCESS [2021-14-07 04:53:00]    &gt;&gt; &#39;Achl_ACHLO041-09_1_F&#39; is created (Forward Read; FASTA).
SUCCESS [2021-14-07 04:53:00] --------------------------------------------------------
SUCCESS [2021-14-07 04:53:00] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-14-07 04:53:00] --------------------------------------------------------
SUCCESS [2021-14-07 04:53:00]    &gt;&gt; &#39;Achl_ACHLO041-09_2_R&#39; is created (Reverse Read; FASTA).
INFO [2021-14-07 04:53:00]    &gt;&gt; The number of reads detected: 2
Assessing frameshifts in nucleotide sequences:
|=================================================================================================| 100%

Time difference of 0.04 secs
SUCCESS [2021-14-07 04:53:02] ==========================================================
SUCCESS [2021-14-07 04:53:02] ======== &#39;SangerContig&#39; S4 instance is created !! ========
SUCCESS [2021-14-07 04:53:02] ==========================================================
INFO [2021-14-07 04:53:02]    &gt;&gt; 2 read(s) created from FASTA file.
INFO [2021-14-07 04:53:02]      &gt;&gt; 1 reads assigned to &#39;forward reads&#39; according to &#39;csv file&#39;.
INFO [2021-14-07 04:53:02]      &gt;&gt; 1 reads assigned to &#39;reverse reads&#39; according to &#39;csv file&#39;.
INFO [2021-14-07 04:53:02] Aligning consensus reads ...
SUCCESS [2021-14-07 04:53:03] #############################################################
SUCCESS [2021-14-07 04:53:03] ######## &#39;SangerAlignment&#39; S4 instance is created !! ########
SUCCESS [2021-14-07 04:53:03] #############################################################
INFO [2021-14-07 04:53:03]   &gt;&gt; 4 contigs detected from &#39;csv file&#39;.
INFO [2021-14-07 04:53:03]       &gt;&gt; Contig &#39;Achl_ACHLO006-09&#39;:
INFO [2021-14-07 04:53:03]           &gt;&gt; 1 forward reads.
INFO [2021-14-07 04:53:03]           &gt;&gt; 1 reverse reads.
INFO [2021-14-07 04:53:03]       &gt;&gt; Contig &#39;Achl_ACHLO007-09&#39;:
INFO [2021-14-07 04:53:03]           &gt;&gt; 1 forward reads.
INFO [2021-14-07 04:53:03]           &gt;&gt; 1 reverse reads.
INFO [2021-14-07 04:53:03]       &gt;&gt; Contig &#39;Achl_ACHLO040-09&#39;:
INFO [2021-14-07 04:53:03]           &gt;&gt; 1 forward reads.
INFO [2021-14-07 04:53:03]           &gt;&gt; 1 reverse reads.
INFO [2021-14-07 04:53:03]       &gt;&gt; Contig &#39;Achl_ACHLO041-09&#39;:
INFO [2021-14-07 04:53:03]           &gt;&gt; 1 forward reads.
INFO [2021-14-07 04:53:03]           &gt;&gt; 1 reverse reads.
INFO [2021-14-07 04:53:03]   &gt;&gt; 8 reads created from FASTA file.
DEBUG [2021-14-07 04:53:03]    &gt;&gt; For more information, please run &#39;object&#39;.
DEBUG [2021-14-07 04:53:03]    &gt;&gt; Run &#39;object@objectResults@readResultTable&#39; to check the results of each Sanger reads
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
</div>
<div class="section" id="id3">
<h3>(3) Writing <em>SangerAlignment</em> FASTA files <sub>(FASTA)</sub><a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">writeFasta</span><span class="p">(</span><span class="n">my_sangerAlignmentFa</span><span class="p">)</span>
</pre></div>
</div>
<div class="toggle docutils container">
<div class="header docutils container">
<p>Following is the R shell output that you will get.</p>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">14</span><span class="o">-</span><span class="mi">07</span> <span class="mi">04</span><span class="p">:</span><span class="mi">53</span><span class="p">:</span><span class="mi">55</span><span class="p">]</span> <span class="n">Your</span> <span class="nb">input</span> <span class="ow">is</span> <span class="s1">&#39;SangerAlignment&#39;</span> <span class="n">S4</span> <span class="n">instance</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">14</span><span class="o">-</span><span class="mi">07</span> <span class="mi">04</span><span class="p">:</span><span class="mi">53</span><span class="p">:</span><span class="mi">55</span><span class="p">]</span> <span class="o">&gt;&gt;&gt;</span> <span class="n">outputDir</span> <span class="p">:</span> <span class="o">/</span><span class="n">private</span><span class="o">/</span><span class="n">var</span><span class="o">/</span><span class="n">folders</span><span class="o">/</span><span class="mi">33</span><span class="o">/</span><span class="mi">7</span><span class="n">v38jdjd2874jcxb6l71m00h0000gn</span><span class="o">/</span><span class="n">T</span><span class="o">/</span><span class="n">RtmpVyEaqi</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">14</span><span class="o">-</span><span class="mi">07</span> <span class="mi">04</span><span class="p">:</span><span class="mi">53</span><span class="p">:</span><span class="mi">55</span><span class="p">]</span> <span class="n">Start</span> <span class="n">to</span> <span class="n">write</span> <span class="s1">&#39;SangerAlignment&#39;</span> <span class="n">to</span> <span class="n">FASTA</span> <span class="nb">format</span> <span class="o">...</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">14</span><span class="o">-</span><span class="mi">07</span> <span class="mi">04</span><span class="p">:</span><span class="mi">53</span><span class="p">:</span><span class="mi">55</span><span class="p">]</span> <span class="o">&gt;&gt;</span> <span class="n">Writing</span> <span class="s1">&#39;alignment&#39;</span> <span class="n">to</span> <span class="n">FASTA</span> <span class="o">...</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">14</span><span class="o">-</span><span class="mi">07</span> <span class="mi">04</span><span class="p">:</span><span class="mi">53</span><span class="p">:</span><span class="mi">55</span><span class="p">]</span> <span class="o">&gt;&gt;</span> <span class="n">Writing</span> <span class="s1">&#39;contigs&#39;</span> <span class="n">to</span> <span class="n">FASTA</span> <span class="o">...</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">14</span><span class="o">-</span><span class="mi">07</span> <span class="mi">04</span><span class="p">:</span><span class="mi">53</span><span class="p">:</span><span class="mi">55</span><span class="p">]</span> <span class="o">&gt;&gt;</span> <span class="n">Writing</span> <span class="nb">all</span> <span class="n">single</span> <span class="n">reads</span> <span class="n">to</span> <span class="n">FASTA</span> <span class="o">...</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">14</span><span class="o">-</span><span class="mi">07</span> <span class="mi">04</span><span class="p">:</span><span class="mi">53</span><span class="p">:</span><span class="mi">55</span><span class="p">]</span> <span class="n">Finish</span> <span class="n">writing</span> <span class="s1">&#39;SangerAlignment&#39;</span> <span class="n">to</span> <span class="n">FASTA</span> <span class="nb">format</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<p>You will get three FASTA files:</p>
<ol class="arabic simple">
<li><p><a class="reference download internal" download="" href="../_downloads/165af37e8893cfdfa4dba91c44cd791b/Sanger_contigs_unalignment.fa"><code class="xref download docutils literal notranslate"><span class="pre">Sanger_contigs_unalignment.fa</span></code></a></p></li>
<li><p><a class="reference download internal" download="" href="../_downloads/144d09562ae5b2c7dc079e4bc2b33fb7/Sanger_contigs_alignment.fa"><code class="xref download docutils literal notranslate"><span class="pre">Sanger_contigs_alignment.fa</span></code></a></p></li>
<li><p><a class="reference download internal" download="" href="../_downloads/81a7338101f110748cb72056d9975225/Sanger_all_trimmed_reads.fa"><code class="xref download docutils literal notranslate"><span class="pre">Sanger_all_trimmed_reads.fa</span></code></a></p></li>
</ol>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="id4">
<h3>(4) Generating <em>SangerAlignment</em> report <sub>(FASTA)</sub><a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">generateReport</span><span class="p">(</span><span class="n">my_sangerAlignmentFa</span><span class="p">)</span>
</pre></div>
</div>
<p>You can check the html report of <a class="reference external" href="https://kuanhao-chao.github.io/sangeranalyseR_report/SangerAlignment/FASTA/SangerAlignment/SangerAlignment_Report.html">this SangerAlignment example (FASTA)</a>.</p>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="how_to_page.html" class="btn btn-neutral float-right" title="Q &amp; A ..." accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="advance_user_SangerContig_fasta.html" class="btn btn-neutral float-left" title="Advanced User Guide - SangerContig (FASTA)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2019, Kuan-Hao Chao.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 <script type="text/javascript">
    $(document).ready(function() {
        $(".toggle > *").hide();
        $(".toggle .header").show();
        $(".toggle .header").click(function() {
            $(this).parent().children().not(".header").toggle(400);
            $(this).parent().children(".header").toggleClass("open");
        })
    });
</script>


</body>
</html>